{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa841d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "============================================================\n",
      "GROUP: Names & Taxonomy\n",
      "============================================================\n",
      "Number of fields: 14\n",
      "\n",
      "accession                      | Entry\n",
      "id                             | Entry Name\n",
      "gene_names                     | Gene Names\n",
      "gene_oln                       | Gene Names (ordered locus)\n",
      "gene_orf                       | Gene Names (ORF)\n",
      "gene_primary                   | Gene Names (primary)\n",
      "gene_synonym                   | Gene Names (synonym)\n",
      "organism_name                  | Organism\n",
      "organism_id                    | Organism (ID)\n",
      "protein_name                   | Protein names\n",
      "xref_proteomes                 | Proteomes\n",
      "lineage                        | Taxonomic lineage\n",
      "lineage_ids                    | Taxonomic lineage (Ids)\n",
      "virus_hosts                    | Virus hosts\n",
      "\n",
      "============================================================\n",
      "GROUP: Sequences\n",
      "============================================================\n",
      "Number of fields: 19\n",
      "\n",
      "cc_alternative_products        | Alternative products (isoforms)\n",
      "ft_var_seq                     | Alternative sequence\n",
      "error_gmodel_pred              | Erroneous gene model prediction\n",
      "fragment                       | Fragment\n",
      "organelle                      | Gene encoded by\n",
      "length                         | Length\n",
      "mass                           | Mass\n",
      "cc_mass_spectrometry           | Mass spectrometry\n",
      "ft_variant                     | Natural variant\n",
      "ft_non_cons                    | Non-adjacent residues\n",
      "ft_non_std                     | Non-standard residue\n",
      "ft_non_ter                     | Non-terminal residue\n",
      "cc_polymorphism                | Polymorphism\n",
      "cc_rna_editing                 | RNA Editing\n",
      "sequence                       | Sequence\n",
      "cc_sequence_caution            | Sequence caution\n",
      "ft_conflict                    | Sequence conflict\n",
      "ft_unsure                      | Sequence uncertainty\n",
      "sequence_version               | Sequence version\n",
      "\n",
      "============================================================\n",
      "GROUP: Function\n",
      "============================================================\n",
      "Number of fields: 16\n",
      "\n",
      "absorption                     | Absorption\n",
      "ft_act_site                    | Active site\n",
      "ft_binding                     | Binding site\n",
      "cc_catalytic_activity          | Catalytic activity\n",
      "cc_cofactor                    | Cofactor\n",
      "ft_dna_bind                    | DNA binding\n",
      "ec                             | EC number\n",
      "cc_activity_regulation         | Activity regulation\n",
      "cc_function                    | Function [CC]\n",
      "kinetics                       | Kinetics\n",
      "cc_pathway                     | Pathway\n",
      "ph_dependence                  | pH dependence\n",
      "redox_potential                | Redox potential\n",
      "rhea                           | Rhea ID\n",
      "ft_site                        | Site\n",
      "temp_dependence                | Temperature dependence\n",
      "\n",
      "============================================================\n",
      "GROUP: Miscellaneous\n",
      "============================================================\n",
      "Number of fields: 11\n",
      "\n",
      "annotation_score               | Annotation\n",
      "cc_caution                     | Caution\n",
      "keyword                        | Keywords\n",
      "keywordid                      | Keyword ID\n",
      "cc_miscellaneous               | Miscellaneous [CC]\n",
      "protein_existence              | Protein existence\n",
      "reviewed                       | Reviewed\n",
      "tools                          | Tools\n",
      "uniparc_id                     | UniParc\n",
      "comment_count                  | Comments\n",
      "feature_count                  | Features\n",
      "\n",
      "============================================================\n",
      "GROUP: Interaction\n",
      "============================================================\n",
      "Number of fields: 2\n",
      "\n",
      "cc_interaction                 | Interacts with\n",
      "cc_subunit                     | Subunit structure\n",
      "\n",
      "============================================================\n",
      "GROUP: Expression\n",
      "============================================================\n",
      "Number of fields: 3\n",
      "\n",
      "cc_developmental_stage         | Developmental stage\n",
      "cc_induction                   | Induction\n",
      "cc_tissue_specificity          | Tissue specificity\n",
      "\n",
      "============================================================\n",
      "GROUP: Gene Ontology (GO)\n",
      "============================================================\n",
      "Number of fields: 5\n",
      "\n",
      "go_p                           | Gene Ontology (biological process)\n",
      "go_c                           | Gene Ontology (cellular component)\n",
      "go                             | Gene Ontology (GO)\n",
      "go_f                           | Gene Ontology (molecular function)\n",
      "go_id                          | Gene Ontology IDs\n",
      "\n",
      "============================================================\n",
      "GROUP: Pathology & Biotech\n",
      "============================================================\n",
      "Number of fields: 7\n",
      "\n",
      "cc_allergen                    | Allergenic Properties\n",
      "cc_biotechnology               | Biotechnological use\n",
      "cc_disruption_phenotype        | Disruption phenotype\n",
      "cc_disease                     | Involvement in disease\n",
      "ft_mutagen                     | Mutagenesis\n",
      "cc_pharmaceutical              | Pharmaceutical use\n",
      "cc_toxic_dose                  | Toxic dose\n",
      "\n",
      "============================================================\n",
      "GROUP: Subcellular location\n",
      "============================================================\n",
      "Number of fields: 4\n",
      "\n",
      "ft_intramem                    | Intramembrane\n",
      "cc_subcellular_location        | Subcellular location [CC]\n",
      "ft_topo_dom                    | Topological domain\n",
      "ft_transmem                    | Transmembrane\n",
      "\n",
      "============================================================\n",
      "GROUP: PTM / Processing\n",
      "============================================================\n",
      "Number of fields: 12\n",
      "\n",
      "ft_chain                       | Chain\n",
      "ft_crosslnk                    | Cross-link\n",
      "ft_disulfid                    | Disulfide bond\n",
      "ft_carbohyd                    | Glycosylation\n",
      "ft_init_met                    | Initiator methionine\n",
      "ft_lipid                       | Lipidation\n",
      "ft_mod_res                     | Modified residue\n",
      "ft_peptide                     | Peptide\n",
      "cc_ptm                         | Post-translational modification\n",
      "ft_propep                      | Propeptide\n",
      "ft_signal                      | Signal peptide\n",
      "ft_transit                     | Transit peptide\n",
      "\n",
      "============================================================\n",
      "GROUP: Structure\n",
      "============================================================\n",
      "Number of fields: 4\n",
      "\n",
      "structure_3d                   | 3D\n",
      "ft_strand                      | Beta strand\n",
      "ft_helix                       | Helix\n",
      "ft_turn                        | Turn\n",
      "\n",
      "============================================================\n",
      "GROUP: Publications\n",
      "============================================================\n",
      "Number of fields: 2\n",
      "\n",
      "lit_pubmed_id                  | PubMed ID\n",
      "lit_doi_id                     | DOI ID\n",
      "\n",
      "============================================================\n",
      "GROUP: Date of\n",
      "============================================================\n",
      "Number of fields: 4\n",
      "\n",
      "date_created                   | Date of creation\n",
      "date_modified                  | Date of last modification\n",
      "date_sequence_modified         | Date of last sequence modification\n",
      "version                        | Entry version\n",
      "\n",
      "============================================================\n",
      "GROUP: Family & Domains\n",
      "============================================================\n",
      "Number of fields: 10\n",
      "\n",
      "ft_coiled                      | Coiled coil\n",
      "ft_compbias                    | Compositional bias\n",
      "cc_domain                      | Domain [CC]\n",
      "ft_domain                      | Domain [FT]\n",
      "ft_motif                       | Motif\n",
      "protein_families               | Protein families\n",
      "ft_region                      | Region\n",
      "ft_repeat                      | Repeat\n",
      "cc_similarity                  | Sequence similarities\n",
      "ft_zn_fing                     | Zinc finger\n",
      "\n",
      "============================================================\n",
      "GROUP: Sequence\n",
      "============================================================\n",
      "Number of fields: 5\n",
      "\n",
      "xref_ccds                      | CCDS\n",
      "xref_embl                      | EMBL\n",
      "xref_generif                   | GeneRIF\n",
      "xref_pir                       | PIR\n",
      "xref_refseq                    | RefSeq\n",
      "\n",
      "============================================================\n",
      "GROUP: 3D structure\n",
      "============================================================\n",
      "Number of fields: 8\n",
      "\n",
      "xref_alphafolddb               | AlphaFoldDB\n",
      "xref_bmrb                      | BMRB\n",
      "xref_emdb                      | EMDB\n",
      "xref_pcddb                     | PCDDB\n",
      "xref_pdb                       | PDB\n",
      "xref_pdbsum                    | PDBsum\n",
      "xref_sasbdb                    | SASBDB\n",
      "xref_smr                       | SMR\n",
      "\n",
      "============================================================\n",
      "GROUP: Protein-protein interaction\n",
      "============================================================\n",
      "Number of fields: 9\n",
      "\n",
      "xref_biogrid                   | BioGRID\n",
      "xref_corum                     | CORUM\n",
      "xref_complexportal             | ComplexPortal\n",
      "xref_dip                       | DIP\n",
      "xref_elm                       | ELM\n",
      "xref_funcoup                   | FunCoup\n",
      "xref_intact                    | IntAct\n",
      "xref_mint                      | MINT\n",
      "xref_string                    | STRING\n",
      "\n",
      "============================================================\n",
      "GROUP: Chemistry\n",
      "============================================================\n",
      "Number of fields: 6\n",
      "\n",
      "xref_bindingdb                 | BindingDB\n",
      "xref_chembl                    | ChEMBL\n",
      "xref_drugbank                  | DrugBank\n",
      "xref_drugcentral               | DrugCentral\n",
      "xref_guidetopharmacology       | GuidetoPHARMACOLOGY\n",
      "xref_swisslipids               | SwissLipids\n",
      "\n",
      "============================================================\n",
      "GROUP: Protein family/group\n",
      "============================================================\n",
      "Number of fields: 12\n",
      "\n",
      "xref_allergome                 | Allergome\n",
      "xref_card                      | CARD\n",
      "xref_cazy                      | CAZy\n",
      "xref_esther                    | ESTHER\n",
      "xref_imgt_gene-db              | IMGT_GENE-DB\n",
      "xref_merops                    | MEROPS\n",
      "xref_moondb                    | MoonDB\n",
      "xref_moonprot                  | MoonProt\n",
      "xref_peroxibase                | PeroxiBase\n",
      "xref_rebase                    | REBASE\n",
      "xref_tcdb                      | TCDB\n",
      "xref_unilectin                 | UniLectin\n",
      "\n",
      "============================================================\n",
      "GROUP: PTM\n",
      "============================================================\n",
      "Number of fields: 10\n",
      "\n",
      "xref_carbonyldb                | CarbonylDB\n",
      "xref_depod                     | DEPOD\n",
      "xref_glyconnect                | GlyConnect\n",
      "xref_glycosmos                 | GlyCosmos\n",
      "xref_glygen                    | GlyGen\n",
      "xref_metosite                  | MetOSite\n",
      "xref_phosphositeplus           | PhosphoSitePlus\n",
      "xref_swisspalm                 | SwissPalm\n",
      "xref_unicarbkb                 | UniCarbKB\n",
      "xref_iptmnet                   | iPTMnet\n",
      "\n",
      "============================================================\n",
      "GROUP: Genetic variation\n",
      "============================================================\n",
      "Number of fields: 4\n",
      "\n",
      "xref_alzforum                  | Alzforum\n",
      "xref_biomuta                   | BioMuta\n",
      "xref_dmdm                      | DMDM\n",
      "xref_dbsnp                     | dbSNP\n",
      "\n",
      "============================================================\n",
      "GROUP: 2D gel\n",
      "============================================================\n",
      "Number of fields: 2\n",
      "\n",
      "xref_ogp                       | OGP\n",
      "xref_reproduction-2dpage       | REPRODUCTION-2DPAGE\n",
      "\n",
      "============================================================\n",
      "GROUP: Proteomic\n",
      "============================================================\n",
      "Number of fields: 10\n",
      "\n",
      "xref_cptac                     | CPTAC\n",
      "xref_massive                   | MassIVE\n",
      "xref_pride                     | PRIDE\n",
      "xref_paxdb                     | PaxDb\n",
      "xref_peptideatlas              | PeptideAtlas\n",
      "xref_promex                    | ProMEX\n",
      "xref_proteomicsdb              | ProteomicsDB\n",
      "xref_pumba                     | Pumba\n",
      "xref_topdownproteomics         | TopDownProteomics\n",
      "xref_jpost                     | jPOST\n",
      "\n",
      "============================================================\n",
      "GROUP: Protocols and materials\n",
      "============================================================\n",
      "Number of fields: 5\n",
      "\n",
      "xref_abcd                      | ABCD\n",
      "xref_antibodypedia             | Antibodypedia\n",
      "xref_cptc                      | CPTC\n",
      "xref_dnasu                     | DNASU\n",
      "xref_ycharos                   | YCharOS\n",
      "\n",
      "============================================================\n",
      "GROUP: Genome annotation\n",
      "============================================================\n",
      "Number of fields: 14\n",
      "\n",
      "xref_ensembl                   | Ensembl\n",
      "xref_ensemblbacteria           | EnsemblBacteria\n",
      "xref_ensemblfungi              | EnsemblFungi\n",
      "xref_ensemblmetazoa            | EnsemblMetazoa\n",
      "xref_ensemblplants             | EnsemblPlants\n",
      "xref_ensemblprotists           | EnsemblProtists\n",
      "xref_geneid                    | GeneID\n",
      "xref_gramene                   | Gramene\n",
      "xref_kegg                      | KEGG\n",
      "xref_mane-select               | MANE-Select\n",
      "xref_patric                    | PATRIC\n",
      "xref_ucsc                      | UCSC\n",
      "xref_vectorbase                | VectorBase\n",
      "xref_wbparasite                | WBParaSite\n",
      "\n",
      "============================================================\n",
      "GROUP: Organism-specific\n",
      "============================================================\n",
      "Number of fields: 39\n",
      "\n",
      "xref_agr                       | AGR\n",
      "xref_arachnoserver             | ArachnoServer\n",
      "xref_araport                   | Araport\n",
      "xref_cgd                       | CGD\n",
      "xref_ctd                       | CTD\n",
      "xref_conoserver                | ConoServer\n",
      "xref_disgenet                  | DisGeNET\n",
      "xref_echobase                  | EchoBASE\n",
      "xref_flybase                   | FlyBase\n",
      "xref_genecards                 | GeneCards\n",
      "xref_genereviews               | GeneReviews\n",
      "xref_hgnc                      | HGNC\n",
      "xref_hpa                       | HPA\n",
      "xref_ic4r                      | IC4R\n",
      "xref_japonicusdb               | JaponicusDB\n",
      "xref_legiolist                 | LegioList\n",
      "xref_leproma                   | Leproma\n",
      "xref_mgi                       | MGI\n",
      "xref_mim                       | MIM\n",
      "xref_maizegdb                  | MaizeGDB\n",
      "xref_malacards                 | MalaCards\n",
      "xref_niagads                   | NIAGADS\n",
      "xref_opentargets               | OpenTargets\n",
      "xref_orphanet                  | Orphanet\n",
      "xref_pharmgkb                  | PharmGKB\n",
      "xref_pombase                   | PomBase\n",
      "xref_pseudocap                 | PseudoCAP\n",
      "xref_rgd                       | RGD\n",
      "xref_sgd                       | SGD\n",
      "xref_tair                      | TAIR\n",
      "xref_tuberculist               | TubercuList\n",
      "xref_veupathdb                 | VEuPathDB\n",
      "xref_vgnc                      | VGNC\n",
      "xref_wormbase                  | WormBase\n",
      "xref_xenbase                   | Xenbase\n",
      "xref_zfin                      | ZFIN\n",
      "xref_dictybase                 | dictyBase\n",
      "xref_euhcvdb                   | euHCVdb\n",
      "xref_nextprot                  | neXtProt\n",
      "\n",
      "============================================================\n",
      "GROUP: Phylogenomic\n",
      "============================================================\n",
      "Number of fields: 9\n",
      "\n",
      "xref_genetree                  | GeneTree\n",
      "xref_hogenom                   | HOGENOM\n",
      "xref_inparanoid                | InParanoid\n",
      "xref_oma                       | OMA\n",
      "xref_orthodb                   | OrthoDB\n",
      "xref_pan-go                    | PAN-GO\n",
      "xref_phylomedb                 | PhylomeDB\n",
      "xref_treefam                   | TreeFam\n",
      "xref_eggnog                    | eggNOG\n",
      "\n",
      "============================================================\n",
      "GROUP: Enzyme and pathway\n",
      "============================================================\n",
      "Number of fields: 10\n",
      "\n",
      "xref_brenda                    | BRENDA\n",
      "xref_biocyc                    | BioCyc\n",
      "xref_pathwaycommons            | PathwayCommons\n",
      "xref_plantreactome             | PlantReactome\n",
      "xref_reactome                  | Reactome\n",
      "xref_sabio-rk                  | SABIO-RK\n",
      "xref_signor                    | SIGNOR\n",
      "xref_strenda-db                | STRENDA-DB\n",
      "xref_signalink                 | SignaLink\n",
      "xref_unipathway                | UniPathway\n",
      "\n",
      "============================================================\n",
      "GROUP: Miscellaneous\n",
      "============================================================\n",
      "Number of fields: 14\n",
      "\n",
      "xref_biogrid-orcs              | BioGRID-ORCS\n",
      "xref_cd-code                   | CD-CODE\n",
      "xref_chitars                   | ChiTaRS\n",
      "xref_evolutionarytrace         | EvolutionaryTrace\n",
      "xref_genewiki                  | GeneWiki\n",
      "xref_genomernai                | GenomeRNAi\n",
      "xref_orcid                     | ORCID\n",
      "xref_pgenn                     | PGenN\n",
      "xref_phi-base                  | PHI-base\n",
      "xref_pro                       | PRO\n",
      "xref_pharos                    | Pharos\n",
      "xref_pubtator                  | PubTator\n",
      "xref_rnact                     | RNAct\n",
      "xref_emind                     | eMIND\n",
      "\n",
      "============================================================\n",
      "GROUP: Gene expression\n",
      "============================================================\n",
      "Number of fields: 4\n",
      "\n",
      "xref_bgee                      | Bgee\n",
      "xref_cleanex                   | CleanEx\n",
      "xref_collectf                  | CollecTF\n",
      "xref_expressionatlas           | ExpressionAtlas\n",
      "\n",
      "============================================================\n",
      "GROUP: Family and domain\n",
      "============================================================\n",
      "Number of fields: 17\n",
      "\n",
      "xref_antifam                   | AntiFam\n",
      "xref_cdd                       | CDD\n",
      "xref_disprot                   | DisProt\n",
      "xref_funfam                    | FunFam\n",
      "xref_gene3d                    | Gene3D\n",
      "xref_hamap                     | HAMAP\n",
      "xref_ideal                     | IDEAL\n",
      "xref_interpro                  | InterPro\n",
      "xref_ncbifam                   | NCBIfam\n",
      "xref_panther                   | PANTHER\n",
      "xref_pirsf                     | PIRSF\n",
      "xref_prints                    | PRINTS\n",
      "xref_prosite                   | PROSITE\n",
      "xref_pfam                      | Pfam\n",
      "xref_sfld                      | SFLD\n",
      "xref_smart                     | SMART\n",
      "xref_supfam                    | SUPFAM\n",
      "\n",
      "============================================================\n",
      "TOTAL AVAILABLE FIELDS: 291\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "def show_all_uniprot_fields():\n",
    "    \"\"\"showing ALL available UniProt fields organized by group\"\"\"\n",
    "    url = \"https://rest.uniprot.org/configure/uniprotkb/result-fields\"\n",
    "    response = requests.get(url)\n",
    "    \n",
    "    if response.status_code == 200:\n",
    "        groups = response.json()\n",
    "        \n",
    "        total_fields = 0\n",
    "        \n",
    "        for group in groups:\n",
    "            group_name = group['groupName']\n",
    "            fields = group['fields']\n",
    "            \n",
    "            print(f\"\\n{'='*60}\")\n",
    "            print(f\"GROUP: {group_name}\")\n",
    "            print(f\"{'='*60}\")\n",
    "            print(f\"Number of fields: {len(fields)}\")\n",
    "            print()\n",
    "            \n",
    "            for field in fields:\n",
    "                field_name = field['name']\n",
    "                field_label = field['label']\n",
    "                print(f\"{field_name:<30} | {field_label}\")\n",
    "                total_fields += 1\n",
    "            \n",
    "        print(f\"\\n{'='*60}\")\n",
    "        print(f\"TOTAL AVAILABLE FIELDS: {total_fields}\")\n",
    "        print(f\"{'='*60}\")\n",
    "        \n",
    "        return groups\n",
    "    else:\n",
    "        print(f\"Error: {response.status_code}\")\n",
    "        return []\n",
    "\n",
    "# show everything\n",
    "all_groups = show_all_uniprot_fields()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d57875b7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a9cc2f4e",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e25106d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🚀 Starting stream download of ALL Swiss-Prot proteins...\n",
      "📋 Downloading 64 fields\n",
      "✅ Stream connection successful!\n",
      "💾 Saving data to file...\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import time\n",
    "from io import StringIO\n",
    "\n",
    "def download_swissprot_stream():\n",
    "    \"\"\"Download ALL Swiss-Prot proteins using stream endpoint\"\"\"\n",
    "    \n",
    "    # fields:\n",
    "    fields = [\n",
    "        # Basic info (8)\n",
    "        'accession', 'id', 'gene_names', 'organism_name', 'protein_name',\n",
    "        'length', 'mass', 'sequence',\n",
    "\n",
    "        # Function (11)\n",
    "        'ft_act_site', 'ft_binding', 'ft_non_std', 'ft_site', 'ft_dna_bind',\n",
    "        'cc_catalytic_activity', 'cc_function', 'ec', 'cc_pathway',\n",
    "        'cc_cofactor', 'cc_activity_regulation',\n",
    "\n",
    "        # Domains & Families (9)\n",
    "        'ft_domain', 'ft_motif', 'ft_region', 'ft_repeat', 'ft_zn_fing',\n",
    "        'ft_coiled', 'ft_compbias', 'protein_families', 'cc_domain',\n",
    "        \n",
    "        # PTM/Processing (11)\n",
    "        'ft_mod_res', 'ft_carbohyd', 'ft_lipid', 'ft_signal', 'ft_transit',\n",
    "        'ft_disulfid', 'ft_chain', 'cc_ptm', 'ft_crosslnk', 'ft_propep', 'ft_peptide',\n",
    "\n",
    "        # Structure (3)\n",
    "        'ft_helix', 'ft_strand', 'ft_turn',\n",
    "        \n",
    "        # Location (4)\n",
    "        'ft_transmem', 'ft_intramem', 'ft_topo_dom', 'cc_subcellular_location',\n",
    "\n",
    "        # Gene Ontology (4)\n",
    "        'go', 'go_f', 'go_p', 'go_c',\n",
    "\n",
    "        # Cross-references (13)\n",
    "        'keyword', 'xref_interpro', 'xref_pfam', 'xref_smart', 'xref_prosite',\n",
    "        'xref_pdb', 'xref_alphafolddb', 'xref_cdd', 'xref_hamap',\n",
    "        'xref_panther', 'xref_pirsf', 'xref_prints', 'xref_supfam', 'xref_gene3d'\n",
    "    ]\n",
    "    \n",
    "    params = {\n",
    "        'query': 'reviewed:true',  # all Swiss-Prot proteins\n",
    "        'format': 'tsv',           # TSV format (easier to parse)\n",
    "        'fields': ','.join(fields) \n",
    "    }\n",
    "    \n",
    "    print(\"🚀 Starting stream download of ALL Swiss-Prot proteins...\")\n",
    "    print(f\"📋 Downloading {len(fields)} fields\")\n",
    "    \n",
    "    try:\n",
    "        # use stream endpoint, which allows for all proteins to be downloaded at once\n",
    "        response = requests.get(\n",
    "            \"https://rest.uniprot.org/uniprotkb/stream\", \n",
    "            params=params,\n",
    "            timeout=900, \n",
    "            stream=True  \n",
    "        )\n",
    "        \n",
    "        if response.status_code == 200:\n",
    "            print(\"✅ Stream connection successful!\")\n",
    "            print(\"💾 Saving data to file...\")\n",
    "            \n",
    "            # save raw TSV to file first \n",
    "            filename = 'swissprot_full_dataset.tsv'\n",
    "            with open(filename, 'w', encoding='utf-8') as f:\n",
    "                for chunk in response.iter_content(chunk_size=8192, decode_unicode=True):\n",
    "                    if chunk:  \n",
    "                        f.write(chunk)\n",
    "            \n",
    "            print(f\"✅ Raw data saved to {filename}\")\n",
    "            print(\"📊 Loading into pandas DataFrame...\")\n",
    "            \n",
    "            df = pd.read_csv(filename, sep='\\t', low_memory=False)\n",
    "            \n",
    "            print(f\"🎉 SUCCESS!\")\n",
    "            print(f\"📈 Downloaded: {len(df):,} proteins\")\n",
    "            print(f\"📋 Fields: {len(df.columns)} columns\")\n",
    "            print(f\"🌍 Organisms: {df['Organism'].nunique():,} species\")\n",
    "            \n",
    "            # validation\n",
    "            print(f\"\\n🔍 validation:\")\n",
    "            print(f\"   unique accessions: {df['Entry'].nunique():,}\")\n",
    "            print(f\"   unique sequences: {df['Sequence'].nunique():,}\")\n",
    "            print(f\"   avg length: {df['Length'].mean():.0f} amino acids\")\n",
    "            \n",
    "            return df\n",
    "            \n",
    "        else:\n",
    "            print(f\"stream failed with status: {response.status_code}\")\n",
    "            print(f\"response: {response.text[:300]}\")\n",
    "            return None\n",
    "            \n",
    "    except requests.exceptions.Timeout:\n",
    "        print(\"Download timed out after 15 minutes\")\n",
    "        return None\n",
    "    except Exception as e:\n",
    "        print(f\"Error during download: {e}\")\n",
    "        return None\n",
    "\n",
    "# Download ALL Swiss-Prot proteins\n",
    "df_full = download_swissprot_stream()\n",
    "\n",
    "if df_full is not None:\n",
    "    compressed_filename = 'data/swissprot_full_dataset.tsv.gz'\n",
    "    df_full.to_csv(compressed_filename, sep='\\t', index=False, compression='gzip')\n",
    "    print(f\"💾 Compressed dataset saved as: {compressed_filename}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee181826",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_df(df):\n",
    "  \n",
    "  print(\"=== DATASET ANALYSIS ===\")\n",
    "  \n",
    "  # 1. Annotation coverage analysis\n",
    "  print(\"\\n Annotation Coverage:\")\n",
    "  high_coverage_fields = []\n",
    "  coverage_data = []\n",
    "  \n",
    "  for col in df.columns:\n",
    "      if col not in ['Entry', 'Entry Name', 'Sequence']:\n",
    "          coverage = df[col].notna().sum()\n",
    "          percentage = coverage / len(df) * 100\n",
    "          \n",
    "          coverage_data.append({\n",
    "              'field': col,\n",
    "              'coverage': coverage,\n",
    "              'percentage': percentage\n",
    "          })\n",
    "          \n",
    "          if percentage >= 50:\n",
    "              high_coverage_fields.append(col)\n",
    "  \n",
    "  # Sort by percentage (highest to lowest)\n",
    "  coverage_data.sort(key=lambda x: x['percentage'], reverse=True)\n",
    "  \n",
    "  # Print sorted coverage\n",
    "  for item in coverage_data:\n",
    "      col = item['field']\n",
    "      coverage = item['coverage']\n",
    "      percentage = item['percentage']\n",
    "      \n",
    "      # for visual purposes\n",
    "      if percentage >= 80:\n",
    "          indicator = \"🟢\"  \n",
    "      elif percentage >= 50:\n",
    "          indicator = \"🟡\"  \n",
    "      elif percentage >= 10:\n",
    "          indicator = \"🟠\"\n",
    "      else:\n",
    "          indicator = \"🔴\" \n",
    "          \n",
    "      print(f\"{indicator} {col}: {coverage:,}/{len(df):,} ({percentage:.1f}%)\")\n",
    "  \n",
    "  # 2. organsism diversity\n",
    "  print(f\"\\n Organism Diversity:\")\n",
    "  top_organisms = df['Organism'].value_counts().head(10)\n",
    "  print(\"Top 10 organisms:\")\n",
    "  for org, count in top_organisms.items():\n",
    "      print(f\"  {org}: {count:,} proteins\")\n",
    "  \n",
    "  # 3. protein length dist\n",
    "  print(f\"\\n📏 Protein Length Statistics:\")\n",
    "  print(f\"  Min length: {df['Length'].min()} amino acids\")\n",
    "  print(f\"  Max length: {df['Length'].max()} amino acids\")\n",
    "  print(f\"  Median length: {df['Length'].median():.0f} amino acids\")\n",
    "  print(f\"  Proteins >1000 AA: {(df['Length'] > 1000).sum():,}\")\n",
    "  \n",
    "  # 4. d0main analysis\n",
    "  if 'Domain [FT]' in df.columns:\n",
    "      domain_coverage = df['Domain [FT]'].notna().sum()\n",
    "      print(f\"\\n Domain Analysis:\")\n",
    "      print(f\"  Proteins with domains: {domain_coverage:,} ({domain_coverage/len(df)*100:.1f}%)\")\n",
    "  \n",
    "  # 5. Overall coverage summary\n",
    "  print(f\"\\n Coverage Summary:\")\n",
    "  high_cov = sum(1 for item in coverage_data if item['percentage'] >= 80)\n",
    "  medium_cov = sum(1 for item in coverage_data if 50 <= item['percentage'] < 80)\n",
    "  low_cov = sum(1 for item in coverage_data if 10 <= item['percentage'] < 50)\n",
    "  very_low_cov = sum(1 for item in coverage_data if item['percentage'] < 10)\n",
    "  \n",
    "  print(f\"  High coverage (≥80%): {high_cov} fields\")\n",
    "  print(f\"  Medium coverage (50-79%): {medium_cov} fields\") \n",
    "  print(f\"  Low coverage (10-49%): {low_cov} fields\")\n",
    "  print(f\"  Very low coverage (<10%): {very_low_cov} fields\")\n",
    "  \n",
    "  return {\n",
    "      'total_proteins': len(df),\n",
    "      'unique_sequences': df['Sequence'].nunique(),\n",
    "      'organisms': df['Organism'].nunique(),\n",
    "      'high_coverage_fields': high_coverage_fields,\n",
    "      'coverage_data': coverage_data \n",
    "  }\n",
    "\n",
    "analysis = analyze_df(df_full)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd5d4cf9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🚀 Starting stream download of ALL Swiss-Prot proteins...\n",
      "📋 Downloading 259 fields\n",
      "✅ Stream connection successful!\n",
      "💾 Saving data to file...\n",
      "✅ Raw data saved to swissprot_full_dataset.tsv\n",
      "📊 Loading into pandas DataFrame...\n",
      "🎉 SUCCESS!\n",
      "📈 Downloaded: 573,661 proteins\n",
      "📋 Fields: 259 columns\n",
      "🌍 Organisms: 14,803 species\n",
      "\n",
      "🔍 validation:\n",
      "   unique accessions: 573,661\n",
      "   unique sequences: 485,423\n",
      "   avg length: 362 amino acids\n",
      "💾 Compressed dataset saved as: swissprot_all_fields.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "## testing with *almost * all fields:\n",
    "\n",
    "import requests\n",
    "import pandas as pd\n",
    "import time\n",
    "from io import StringIO\n",
    "\n",
    "def download_swissprot_stream():\n",
    "    \"\"\"Download ALL Swiss-Prot proteins using stream endpoint\"\"\"\n",
    "    \n",
    "    # fields:\n",
    "    fields = [\n",
    "        # Basic info (8)\n",
    "        'accession', 'id', 'gene_names', 'organism_name', 'protein_name',\n",
    "        'length', 'mass', 'sequence', \n",
    "        # additional fields:\n",
    "        'gene_oln', 'gene_orf', 'gene_primary', 'gene_synonym', 'lineage', 'lineage_ids', 'virus_hosts',\n",
    "\n",
    "        # Function (11)\n",
    "        'ft_act_site', 'ft_binding', 'ft_non_std', 'ft_site', 'ft_dna_bind',\n",
    "        'cc_catalytic_activity', 'cc_function', 'ec', 'cc_pathway',\n",
    "        'cc_cofactor', 'cc_activity_regulation',\n",
    "        # additional fields:\n",
    "        'temp_dependence', 'redox_potential', 'ph_dependence', 'kinetics',\n",
    "\n",
    "        # additional fields (interaction + expression):\n",
    "        'cc_interaction', 'cc_subunit', 'cc_developmental_stage', 'cc_induction', 'cc_tissue_specificity',\n",
    "\n",
    "        # Domains & Families (9)\n",
    "        'ft_domain', 'ft_motif', 'ft_region', 'ft_repeat', 'ft_zn_fing',\n",
    "        'ft_coiled', 'ft_compbias', 'protein_families', 'cc_domain',\n",
    "        \n",
    "        # PTM/Processing (11)\n",
    "        'ft_mod_res', 'ft_carbohyd', 'ft_lipid', 'ft_signal', 'ft_transit',\n",
    "        'ft_disulfid', 'ft_chain', 'cc_ptm', 'ft_crosslnk', 'ft_propep', 'ft_peptide',\n",
    "\n",
    "        # Structure (3)\n",
    "        'ft_helix', 'ft_strand', 'ft_turn',\n",
    "        \n",
    "        # Location (4)\n",
    "        'ft_transmem', 'ft_intramem', 'ft_topo_dom', 'cc_subcellular_location',\n",
    "\n",
    "        # Gene Ontology (4)\n",
    "        'go', 'go_f', 'go_p', 'go_c',\n",
    "\n",
    "        # Cross-references (13)\n",
    "        'keyword', 'xref_interpro', 'xref_pfam', 'xref_smart', 'xref_prosite',\n",
    "        'xref_pdb', 'xref_alphafolddb', 'xref_cdd', 'xref_hamap',\n",
    "        'xref_panther', 'xref_pirsf', 'xref_prints', 'xref_supfam', 'xref_gene3d',\n",
    "\n",
    "        # extra references:\n",
    "        'xref_ccds', \n",
    "        'xref_embl', \n",
    "        'xref_generif', \n",
    "        'xref_pir',\n",
    "        'xref_refseq',\n",
    "        'xref_bmrb',\n",
    "        'xref_emdb', \n",
    "        'xref_pcddb',\n",
    "        'xref_pdbsum', \n",
    "        'xref_sasbdb', \n",
    "        'xref_smr',\n",
    "        'xref_antifam',\n",
    "        'xref_cdd',\n",
    "        'xref_disprot',\n",
    "        'xref_funfam',\n",
    "        'xref_gene3d',\n",
    "        'xref_hamap',\n",
    "        'xref_ideal',\n",
    "        'xref_interpro',\n",
    "        'xref_ncbifam',\n",
    "        'xref_panther',\n",
    "        'xref_pirsf',\n",
    "        'xref_prints',\n",
    "        'xref_prosite', \n",
    "        'xref_pfam', \n",
    "        'xref_sfld', \n",
    "        'xref_smart', \n",
    "        'xref_supfam', \n",
    "        'xref_bgee',\n",
    "        'xref_cleanex',\n",
    "        'xref_collectf',\n",
    "        'xref_expressionatlas',\n",
    "        'xref_biogrid-orcs',\n",
    "        'xref_cd-code',\n",
    "        'xref_chitars',\n",
    "        'xref_evolutionarytrace',\n",
    "        'xref_genewiki',\n",
    "        'xref_genomernai',\n",
    "        'xref_orcid',\n",
    "        'xref_pgenn',\n",
    "        'xref_phi-base',\n",
    "        'xref_pro',\n",
    "        'xref_pharos',\n",
    "        'xref_pubtator',\n",
    "        'xref_rnact',\n",
    "        'xref_emind',\n",
    "        'xref_brenda',\n",
    "        'xref_biocyc',\n",
    "        'xref_pathwaycommons',\n",
    "        'xref_plantreactome',\n",
    "        'xref_reactome',\n",
    "        'xref_sabio-rk',\n",
    "        'xref_signor',\n",
    "        'xref_strenda-db',\n",
    "        'xref_signalink',\n",
    "        'xref_unipathway',\n",
    "        'xref_genetree',\n",
    "        'xref_hogenom',\n",
    "        'xref_inparanoid',\n",
    "        'xref_oma',\n",
    "        'xref_orthodb',\n",
    "        'xref_pan-go',\n",
    "        'xref_phylomedb',\n",
    "        'xref_treefam',\n",
    "        'xref_eggnog',\n",
    "        'xref_agr',\n",
    "        'xref_arachnoserver',\n",
    "        'xref_araport',\n",
    "        'xref_cgd',\n",
    "        'xref_ctd',\n",
    "        'xref_conoserver',\n",
    "        'xref_disgenet',\n",
    "        'xref_echobase',\n",
    "        'xref_flybase',\n",
    "        'xref_genecards',\n",
    "        'xref_genereviews',\n",
    "        'xref_hgnc',\n",
    "        'xref_hpa',\n",
    "        'xref_ic4r',\n",
    "        'xref_japonicusdb',\n",
    "        'xref_legiolist',\n",
    "        'xref_leproma',\n",
    "        'xref_mgi',\n",
    "        'xref_mim',\n",
    "        'xref_maizegdb',\n",
    "        'xref_malacards',\n",
    "        'xref_niagads',\n",
    "        'xref_opentargets',\n",
    "        'xref_orphanet',\n",
    "        'xref_pharmgkb',\n",
    "        'xref_pombase',\n",
    "        'xref_pseudocap',\n",
    "        'xref_rgd',\n",
    "        'xref_sgd',\n",
    "        'xref_tair',\n",
    "        'xref_tuberculist',\n",
    "        'xref_veupathdb',\n",
    "        'xref_vgnc',\n",
    "        'xref_wormbase',\n",
    "        'xref_xenbase',\n",
    "        'xref_zfin',\n",
    "        'xref_dictybase',\n",
    "        'xref_euhcvdb',\n",
    "        'xref_nextprot',\n",
    "        'xref_ensembl',\n",
    "        'xref_ensemblbacteria',\n",
    "        'xref_ensemblfungi',\n",
    "        'xref_ensemblmetazoa',\n",
    "        'xref_ensemblplants',\n",
    "        'xref_ensemblprotists',\n",
    "        'xref_geneid',\n",
    "        'xref_gramene',\n",
    "        'xref_kegg',\n",
    "        'xref_mane-select',\n",
    "        'xref_patric',\n",
    "        'xref_ucsc',\n",
    "        'xref_vectorbase',\n",
    "        'xref_wbparasite',\n",
    "        'xref_abcd',\n",
    "        'xref_antibodypedia',\n",
    "        'xref_cptc',\n",
    "        'xref_dnasu',\n",
    "        'xref_ycharos',\n",
    "        'xref_cptac',\n",
    "        'xref_massive',\n",
    "        'xref_pride',\n",
    "        'xref_paxdb',\n",
    "        'xref_peptideatlas',\n",
    "        'xref_promex',\n",
    "        'xref_proteomicsdb',\n",
    "        'xref_pumba',\n",
    "        'xref_topdownproteomics',\n",
    "        'xref_jpost',\n",
    "        'xref_ogp',\n",
    "        'xref_reproduction-2dpage',\n",
    "        'xref_alzforum',\n",
    "        'xref_biomuta',\n",
    "        'xref_dmdm',\n",
    "        'xref_dbsnp',\n",
    "        'xref_carbonyldb',\n",
    "        'xref_depod',\n",
    "        'xref_glyconnect',\n",
    "        'xref_glycosmos',\n",
    "        'xref_glygen',\n",
    "        'xref_metosite',\n",
    "        'xref_phosphositeplus',\n",
    "        'xref_swisspalm',\n",
    "        'xref_unicarbkb',\n",
    "        'xref_iptmnet',\n",
    "        'xref_allergome',\n",
    "        'xref_card',\n",
    "        'xref_cazy',\n",
    "        'xref_esther',\n",
    "        'xref_imgt_gene-db',\n",
    "        'xref_merops',\n",
    "        'xref_moondb',\n",
    "        'xref_moonprot',\n",
    "        'xref_peroxibase',\n",
    "        'xref_rebase',\n",
    "        'xref_tcdb',\n",
    "        'xref_unilectin',\n",
    "        'xref_bindingdb',\n",
    "        'xref_chembl',\n",
    "        'xref_drugbank',\n",
    "        'xref_drugcentral',\n",
    "        'xref_guidetopharmacology',\n",
    "        'xref_swisslipids',\n",
    "        'xref_biogrid',\n",
    "        'xref_corum',\n",
    "        'xref_complexportal',\n",
    "        'xref_dip',\n",
    "        'xref_elm',\n",
    "        'xref_funcoup',\n",
    "        'xref_intact',\n",
    "        'xref_mint',\n",
    "        'xref_string',\n",
    "        'xref_pdbsum',\n",
    "        'xref_sasbdb',\n",
    "        'xref_smr'\n",
    "    ]\n",
    "    \n",
    "    params = {\n",
    "        'query': 'reviewed:true',  # all Swiss-Prot proteins\n",
    "        'format': 'tsv',           # TSV format (easier to parse)\n",
    "        'fields': ','.join(fields) \n",
    "    }\n",
    "    \n",
    "    print(\"🚀 Starting stream download of ALL Swiss-Prot proteins...\")\n",
    "    print(f\"📋 Downloading {len(fields)} fields\")\n",
    "    \n",
    "    try:\n",
    "        # use stream endpoint, which allows for all proteins to be downloaded at once\n",
    "        response = requests.get(\n",
    "            \"https://rest.uniprot.org/uniprotkb/stream\", \n",
    "            params=params,\n",
    "            timeout=900, \n",
    "            stream=True  \n",
    "        )\n",
    "        \n",
    "        if response.status_code == 200:\n",
    "            print(\"Stream connection successful!\")\n",
    "            print(\"Saving data to file...\")\n",
    "            \n",
    "            # save raw TSV to file first \n",
    "            filename = 'swissprot_full_dataset.tsv'\n",
    "            with open(filename, 'w', encoding='utf-8') as f:\n",
    "                for chunk in response.iter_content(chunk_size=8192, decode_unicode=True):\n",
    "                    if chunk:  \n",
    "                        f.write(chunk)\n",
    "            \n",
    "            print(f\"Raw data saved to {filename}\")\n",
    "            print(\"Loading into pandas DataFrame...\")\n",
    "\n",
    "            df = pd.read_csv(filename, sep='\\t', low_memory=False)\n",
    "            \n",
    "            print(f\"SUCCESS!\")\n",
    "            print(f\"Downloaded: {len(df):,} proteins\")\n",
    "            print(f\"Fields: {len(df.columns)} columns\")\n",
    "            print(f\"Organisms: {df['Organism'].nunique():,} species\")\n",
    "            \n",
    "            # validation\n",
    "            print(f\"\\nvalidation:\")\n",
    "            print(f\"   unique accessions: {df['Entry'].nunique():,}\")\n",
    "            print(f\"   unique sequences: {df['Sequence'].nunique():,}\")\n",
    "            print(f\"   avg length: {df['Length'].mean():.0f} amino acids\")\n",
    "            \n",
    "            return df\n",
    "            \n",
    "        else:\n",
    "            print(f\"stream failed with status: {response.status_code}\")\n",
    "            print(f\"response: {response.text[:300]}\")\n",
    "            return None\n",
    "            \n",
    "    except requests.exceptions.Timeout:\n",
    "        print(\"Download timed out after 15 minutes\")\n",
    "        return None\n",
    "    except Exception as e:\n",
    "        print(f\"Error during download: {e}\")\n",
    "        return None\n",
    "\n",
    "# Download ALL Swiss-Prot proteins\n",
    "df_full = download_swissprot_stream()\n",
    "\n",
    "if df_full is not None:\n",
    "    compressed_filename = 'swissprot_all_fields.tsv.gz'\n",
    "    df_full.to_csv(compressed_filename, sep='\\t', index=False, compression='gzip')\n",
    "    print(f\"💾 Compressed dataset saved as: {compressed_filename}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "4121eff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=== DATASET ANALYSIS ===\n",
      "\n",
      " Annotation Coverage:\n",
      "🟢 Organism: 573,661/573,661 (100.0%)\n",
      "🟢 Protein names: 573,661/573,661 (100.0%)\n",
      "🟢 Length: 573,661/573,661 (100.0%)\n",
      "🟢 Mass: 573,661/573,661 (100.0%)\n",
      "🟢 Taxonomic lineage: 573,661/573,661 (100.0%)\n",
      "🟢 Taxonomic lineage (Ids): 573,661/573,661 (100.0%)\n",
      "🟢 Keywords: 566,515/573,661 (98.8%)\n",
      "🟢 Chain: 565,890/573,661 (98.6%)\n",
      "🟢 EMBL: 560,693/573,661 (97.7%)\n",
      "🟢 InterPro: 555,533/573,661 (96.8%)\n",
      "🟢 InterPro.1: 555,533/573,661 (96.8%)\n",
      "🟢 Gene Ontology (GO): 552,917/573,661 (96.4%)\n",
      "🟢 Gene Names: 548,889/573,661 (95.7%)\n",
      "🟢 AlphaFoldDB: 548,729/573,661 (95.7%)\n",
      "🟢 Pfam: 545,517/573,661 (95.1%)\n",
      "🟢 Pfam.1: 545,517/573,661 (95.1%)\n",
      "🟢 SMR: 524,392/573,661 (91.4%)\n",
      "🟢 SMR.1: 524,392/573,661 (91.4%)\n",
      "🟢 Protein families: 516,427/573,661 (90.0%)\n",
      "🟢 PANTHER: 505,013/573,661 (88.0%)\n",
      "🟢 PANTHER.1: 505,013/573,661 (88.0%)\n",
      "🟢 Gene Names (primary): 502,595/573,661 (87.6%)\n",
      "🟢 Gene Ontology (molecular function): 492,124/573,661 (85.8%)\n",
      "🟢 KEGG: 481,691/573,661 (84.0%)\n",
      "🟢 Gene3D: 477,613/573,661 (83.3%)\n",
      "🟢 Gene3D.1: 477,613/573,661 (83.3%)\n",
      "🟢 Function [CC]: 471,942/573,661 (82.3%)\n",
      "🟢 Gene Ontology (cellular component): 467,660/573,661 (81.5%)\n",
      "🟢 Gene Ontology (biological process): 465,666/573,661 (81.2%)\n",
      "🟢 SUPFAM: 461,267/573,661 (80.4%)\n",
      "🟢 SUPFAM.1: 461,267/573,661 (80.4%)\n",
      "🟡 RefSeq: 447,823/573,661 (78.1%)\n",
      "🟡 HOGENOM: 428,247/573,661 (74.7%)\n",
      "🟡 Gene Names (ordered locus): 380,300/573,661 (66.3%)\n",
      "🟡 Subcellular location [CC]: 360,038/573,661 (62.8%)\n",
      "🟡 NCBIfam: 347,089/573,661 (60.5%)\n",
      "🟡 STRING: 336,721/573,661 (58.7%)\n",
      "🟡 eggNOG: 334,263/573,661 (58.3%)\n",
      "🟡 HAMAP: 328,070/573,661 (57.2%)\n",
      "🟡 HAMAP.1: 328,070/573,661 (57.2%)\n",
      "🟡 FunFam: 327,323/573,661 (57.1%)\n",
      "🟡 PROSITE: 312,414/573,661 (54.5%)\n",
      "🟡 PROSITE.1: 312,414/573,661 (54.5%)\n",
      "🟡 CDD: 309,337/573,661 (53.9%)\n",
      "🟡 CDD.1: 309,337/573,661 (53.9%)\n",
      "🟡 Subunit structure: 296,777/573,661 (51.7%)\n",
      "🟠 EC number: 278,804/573,661 (48.6%)\n",
      "🟠 OrthoDB: 270,086/573,661 (47.1%)\n",
      "🟠 GeneID: 260,629/573,661 (45.4%)\n",
      "🟠 Catalytic activity: 259,184/573,661 (45.2%)\n",
      "🟠 Binding site: 221,442/573,661 (38.6%)\n",
      "🟠 InParanoid: 164,400/573,661 (28.7%)\n",
      "🟠 PaxDb: 154,035/573,661 (26.9%)\n",
      "🟠 Region: 151,443/573,661 (26.4%)\n",
      "🟠 SMART: 149,082/573,661 (26.0%)\n",
      "🟠 SMART.1: 149,082/573,661 (26.0%)\n",
      "🟠 FunCoup: 143,393/573,661 (25.0%)\n",
      "🟠 Domain [FT]: 133,903/573,661 (23.3%)\n",
      "🟠 Pathway: 130,645/573,661 (22.8%)\n",
      "🟠 PRINTS: 129,976/573,661 (22.7%)\n",
      "🟠 PRINTS.1: 129,976/573,661 (22.7%)\n",
      "🟠 UniPathway: 126,687/573,661 (22.1%)\n",
      "🟠 Cofactor: 122,515/573,661 (21.4%)\n",
      "🟠 OMA: 120,581/573,661 (21.0%)\n",
      "🟠 PhylomeDB: 115,758/573,661 (20.2%)\n",
      "🟠 PIR: 114,896/573,661 (20.0%)\n",
      "🟠 PIRSF: 109,989/573,661 (19.2%)\n",
      "🟠 PIRSF.1: 109,989/573,661 (19.2%)\n",
      "🟠 Active site: 106,828/573,661 (18.6%)\n",
      "🟠 PRO: 98,646/573,661 (17.2%)\n",
      "🟠 PATRIC: 93,239/573,661 (16.3%)\n",
      "🟠 Compositional bias: 93,018/573,661 (16.2%)\n",
      "🟠 Gene Names (ORF): 90,890/573,661 (15.8%)\n",
      "🟠 Gene Names (synonym): 89,424/573,661 (15.6%)\n",
      "🟠 Transmembrane: 80,533/573,661 (14.0%)\n",
      "🟠 VEuPathDB: 79,958/573,661 (13.9%)\n",
      "🟠 CTD: 75,470/573,661 (13.2%)\n",
      "🟠 Modified residue: 75,087/573,661 (13.1%)\n",
      "🟠 AGR: 68,532/573,661 (11.9%)\n",
      "🟠 Bgee: 61,882/573,661 (10.8%)\n",
      "🟠 BioGRID: 60,262/573,661 (10.5%)\n",
      "🟠 IntAct: 58,598/573,661 (10.2%)\n",
      "🔴 iPTMnet: 56,779/573,661 (9.9%)\n",
      "🔴 GeneTree: 56,448/573,661 (9.8%)\n",
      "🔴 EnsemblBacteria: 55,357/573,661 (9.6%)\n",
      "🔴 Domain [CC]: 52,671/573,661 (9.2%)\n",
      "🔴 Tissue specificity: 51,344/573,661 (9.0%)\n",
      "🔴 ExpressionAtlas: 51,247/573,661 (8.9%)\n",
      "🔴 Ensembl: 49,662/573,661 (8.7%)\n",
      "🔴 DNASU: 48,433/573,661 (8.4%)\n",
      "🔴 Post-translational modification: 47,178/573,661 (8.2%)\n",
      "🔴 UCSC: 46,564/573,661 (8.1%)\n",
      "🔴 TreeFam: 46,331/573,661 (8.1%)\n",
      "🔴 ProteomicsDB: 45,416/573,661 (7.9%)\n",
      "🔴 Signal peptide: 45,100/573,661 (7.9%)\n",
      "🔴 BioGRID-ORCS: 44,519/573,661 (7.8%)\n",
      "🔴 BioCyc: 44,187/573,661 (7.7%)\n",
      "🔴 RNAct: 43,121/573,661 (7.5%)\n",
      "🔴 PhosphoSitePlus: 42,211/573,661 (7.4%)\n",
      "🔴 Reactome: 39,088/573,661 (6.8%)\n",
      "🔴 PeptideAtlas: 38,903/573,661 (6.8%)\n",
      "🔴 PDB: 37,442/573,661 (6.5%)\n",
      "🔴 PDBsum: 37,442/573,661 (6.5%)\n",
      "🔴 PDBsum.1: 37,442/573,661 (6.5%)\n",
      "🔴 GlyGen: 37,350/573,661 (6.5%)\n",
      "🔴 Disulfide bond: 37,053/573,661 (6.5%)\n",
      "🔴 Site: 36,986/573,661 (6.4%)\n",
      "🔴 CCDS: 34,822/573,661 (6.1%)\n",
      "🔴 Antibodypedia: 32,219/573,661 (5.6%)\n",
      "🔴 Glycosylation: 32,052/573,661 (5.6%)\n",
      "🔴 Motif: 31,683/573,661 (5.5%)\n",
      "🔴 Helix: 31,362/573,661 (5.5%)\n",
      "🔴 Topological domain: 30,888/573,661 (5.4%)\n",
      "🔴 ChiTaRS: 29,758/573,661 (5.2%)\n",
      "🔴 Beta strand: 29,524/573,661 (5.1%)\n",
      "🔴 jPOST: 29,048/573,661 (5.1%)\n",
      "🔴 GlyCosmos: 28,908/573,661 (5.0%)\n",
      "🔴 Induction: 26,498/573,661 (4.6%)\n",
      "🔴 Turn: 25,641/573,661 (4.5%)\n",
      "🔴 Interacts with: 25,016/573,661 (4.4%)\n",
      "🔴 MINT: 24,141/573,661 (4.2%)\n",
      "🔴 EnsemblFungi: 22,968/573,661 (4.0%)\n",
      "🔴 EvolutionaryTrace: 22,685/573,661 (4.0%)\n",
      "🔴 GenomeRNAi: 22,329/573,661 (3.9%)\n",
      "🔴 EnsemblPlants: 21,984/573,661 (3.8%)\n",
      "🔴 Gramene: 21,984/573,661 (3.8%)\n",
      "🔴 neXtProt: 20,298/573,661 (3.5%)\n",
      "🔴 BioMuta: 20,260/573,661 (3.5%)\n",
      "🔴 HGNC: 20,245/573,661 (3.5%)\n",
      "🔴 GeneCards: 20,242/573,661 (3.5%)\n",
      "🔴 PAN-GO: 20,212/573,661 (3.5%)\n",
      "🔴 Pharos: 20,196/573,661 (3.5%)\n",
      "🔴 SignaLink: 19,952/573,661 (3.5%)\n",
      "🔴 PathwayCommons: 19,437/573,661 (3.4%)\n",
      "🔴 HPA: 19,215/573,661 (3.3%)\n",
      "🔴 Activity regulation: 19,155/573,661 (3.3%)\n",
      "🔴 MassIVE: 19,137/573,661 (3.3%)\n",
      "🔴 BRENDA: 18,662/573,661 (3.3%)\n",
      "🔴 MANE-Select: 18,453/573,661 (3.2%)\n",
      "🔴 OpenTargets: 18,424/573,661 (3.2%)\n",
      "🔴 Pumba: 18,204/573,661 (3.2%)\n",
      "🔴 PharmGKB: 18,013/573,661 (3.1%)\n",
      "🔴 DIP: 17,528/573,661 (3.1%)\n",
      "🔴 DisGeNET: 17,412/573,661 (3.0%)\n",
      "🔴 Virus hosts: 17,381/573,661 (3.0%)\n",
      "🔴 MGI: 17,112/573,661 (3.0%)\n",
      "🔴 MIM: 16,424/573,661 (2.9%)\n",
      "🔴 Araport: 16,321/573,661 (2.8%)\n",
      "🔴 TAIR: 16,321/573,661 (2.8%)\n",
      "🔴 DMDM: 16,165/573,661 (2.8%)\n",
      "🔴 Coiled coil: 15,786/573,661 (2.8%)\n",
      "🔴 Repeat: 15,277/573,661 (2.7%)\n",
      "🔴 Developmental stage: 14,621/573,661 (2.5%)\n",
      "🔴 SwissPalm: 13,972/573,661 (2.4%)\n",
      "🔴 MEROPS: 13,836/573,661 (2.4%)\n",
      "🔴 Propeptide: 13,346/573,661 (2.3%)\n",
      "🔴 Zinc finger: 13,190/573,661 (2.3%)\n",
      "🔴 dbSNP: 12,774/573,661 (2.2%)\n",
      "🔴 EnsemblMetazoa: 11,932/573,661 (2.1%)\n",
      "🔴 DNA binding: 10,971/573,661 (1.9%)\n",
      "🔴 EMDB: 10,736/573,661 (1.9%)\n",
      "🔴 GeneWiki: 10,269/573,661 (1.8%)\n",
      "🔴 Kinetics: 9,756/573,661 (1.7%)\n",
      "🔴 Transit peptide: 9,520/573,661 (1.7%)\n",
      "🔴 ComplexPortal: 9,243/573,661 (1.6%)\n",
      "🔴 Cross-link: 9,183/573,661 (1.6%)\n",
      "🔴 SFLD: 9,117/573,661 (1.6%)\n",
      "🔴 ChEMBL: 9,109/573,661 (1.6%)\n",
      "🔴 Lipidation: 8,933/573,661 (1.6%)\n",
      "🔴 Peptide: 8,865/573,661 (1.5%)\n",
      "🔴 CAZy: 8,744/573,661 (1.5%)\n",
      "🔴 TCDB: 8,652/573,661 (1.5%)\n",
      "🔴 CD-CODE: 8,219/573,661 (1.4%)\n",
      "🔴 RGD: 8,151/573,661 (1.4%)\n",
      "🔴 CORUM: 8,089/573,661 (1.4%)\n",
      "🔴 SIGNOR: 7,673/573,661 (1.3%)\n",
      "🔴 MalaCards: 7,260/573,661 (1.3%)\n",
      "🔴 BindingDB: 6,928/573,661 (1.2%)\n",
      "🔴 BMRB: 6,915/573,661 (1.2%)\n",
      "🔴 SGD: 6,748/573,661 (1.2%)\n",
      "🔴 SABIO-RK: 5,951/573,661 (1.0%)\n",
      "🔴 pH dependence: 5,462/573,661 (1.0%)\n",
      "🔴 EnsemblProtists: 5,204/573,661 (0.9%)\n",
      "🔴 PomBase: 5,127/573,661 (0.9%)\n",
      "🔴 WormBase: 5,102/573,661 (0.9%)\n",
      "🔴 DrugBank: 4,937/573,661 (0.9%)\n",
      "🔴 Xenbase: 4,754/573,661 (0.8%)\n",
      "🔴 Orphanet: 4,405/573,661 (0.8%)\n",
      "🔴 EchoBASE: 4,158/573,661 (0.7%)\n",
      "🔴 dictyBase: 4,111/573,661 (0.7%)\n",
      "🔴 FlyBase: 3,867/573,661 (0.7%)\n",
      "🔴 MetOSite: 3,456/573,661 (0.6%)\n",
      "🔴 VGNC: 3,453/573,661 (0.6%)\n",
      "🔴 ZFIN: 3,245/573,661 (0.6%)\n",
      "🔴 ABCD: 3,195/573,661 (0.6%)\n",
      "🔴 Temperature dependence: 3,150/573,661 (0.5%)\n",
      "🔴 ESTHER: 3,022/573,661 (0.5%)\n",
      "🔴 DrugCentral: 2,982/573,661 (0.5%)\n",
      "🔴 TopDownProteomics: 2,956/573,661 (0.5%)\n",
      "🔴 TubercuList: 2,316/573,661 (0.4%)\n",
      "🔴 GuidetoPHARMACOLOGY: 2,278/573,661 (0.4%)\n",
      "🔴 GlyConnect: 2,215/573,661 (0.4%)\n",
      "🔴 CGD: 2,091/573,661 (0.4%)\n",
      "🔴 PseudoCAP: 2,054/573,661 (0.4%)\n",
      "🔴 CPTAC: 1,929/573,661 (0.3%)\n",
      "🔴 PHI-base: 1,918/573,661 (0.3%)\n",
      "🔴 ELM: 1,814/573,661 (0.3%)\n",
      "🔴 DisProt: 1,763/573,661 (0.3%)\n",
      "🔴 GeneReviews: 1,628/573,661 (0.3%)\n",
      "🔴 Intramembrane: 1,497/573,661 (0.3%)\n",
      "🔴 SwissLipids: 1,394/573,661 (0.2%)\n",
      "🔴 Allergome: 1,315/573,661 (0.2%)\n",
      "🔴 CarbonylDB: 1,159/573,661 (0.2%)\n",
      "🔴 ArachnoServer: 1,138/573,661 (0.2%)\n",
      "🔴 IDEAL: 1,101/573,661 (0.2%)\n",
      "🔴 REPRODUCTION-2DPAGE: 1,039/573,661 (0.2%)\n",
      "🔴 SASBDB: 985/573,661 (0.2%)\n",
      "🔴 SASBDB.1: 985/573,661 (0.2%)\n",
      "🔴 ConoServer: 879/573,661 (0.2%)\n",
      "🔴 PeroxiBase: 772/573,661 (0.1%)\n",
      "🔴 PlantReactome: 771/573,661 (0.1%)\n",
      "🔴 LegioList: 763/573,661 (0.1%)\n",
      "🔴 Leproma: 669/573,661 (0.1%)\n",
      "🔴 PRIDE: 637/573,661 (0.1%)\n",
      "🔴 MaizeGDB: 525/573,661 (0.1%)\n",
      "🔴 ProMEX: 489/573,661 (0.1%)\n",
      "🔴 CPTC: 406/573,661 (0.1%)\n",
      "🔴 REBASE: 397/573,661 (0.1%)\n",
      "🔴 OGP: 373/573,661 (0.1%)\n",
      "🔴 MoonProt: 368/573,661 (0.1%)\n",
      "🔴 UniLectin: 367/573,661 (0.1%)\n",
      "🔴 MoonDB: 348/573,661 (0.1%)\n",
      "🔴 CARD: 319/573,661 (0.1%)\n",
      "🔴 Non-standard residue: 285/573,661 (0.0%)\n",
      "🔴 IMGT_GENE-DB: 267/573,661 (0.0%)\n",
      "🔴 DEPOD: 254/573,661 (0.0%)\n",
      "🔴 CollecTF: 138/573,661 (0.0%)\n",
      "🔴 PCDDB: 134/573,661 (0.0%)\n",
      "🔴 Redox potential: 132/573,661 (0.0%)\n",
      "🔴 NIAGADS: 76/573,661 (0.0%)\n",
      "🔴 WBParaSite: 54/573,661 (0.0%)\n",
      "🔴 STRENDA-DB: 45/573,661 (0.0%)\n",
      "🔴 euHCVdb: 44/573,661 (0.0%)\n",
      "🔴 JaponicusDB: 43/573,661 (0.0%)\n",
      "🔴 YCharOS: 36/573,661 (0.0%)\n",
      "🔴 AntiFam: 22/573,661 (0.0%)\n",
      "🔴 GeneRIF: 0/573,661 (0.0%)\n",
      "🔴 CleanEx: 0/573,661 (0.0%)\n",
      "🔴 ORCID: 0/573,661 (0.0%)\n",
      "🔴 PGenN: 0/573,661 (0.0%)\n",
      "🔴 PubTator: 0/573,661 (0.0%)\n",
      "🔴 eMIND: 0/573,661 (0.0%)\n",
      "🔴 IC4R: 0/573,661 (0.0%)\n",
      "🔴 VectorBase: 0/573,661 (0.0%)\n",
      "🔴 Alzforum: 0/573,661 (0.0%)\n",
      "🔴 UniCarbKB: 0/573,661 (0.0%)\n",
      "\n",
      " Organism Diversity:\n",
      "Top 10 organisms:\n",
      "  Homo sapiens (Human): 20,420 proteins\n",
      "  Mus musculus (Mouse): 17,240 proteins\n",
      "  Arabidopsis thaliana (Mouse-ear cress): 16,397 proteins\n",
      "  Rattus norvegicus (Rat): 8,219 proteins\n",
      "  Saccharomyces cerevisiae (strain ATCC 204508 / S288c) (Baker's yeast): 6,733 proteins\n",
      "  Bos taurus (Bovine): 6,052 proteins\n",
      "  Schizosaccharomyces pombe (strain 972 / ATCC 24843) (Fission yeast): 5,123 proteins\n",
      "  Escherichia coli (strain K12): 4,531 proteins\n",
      "  Caenorhabditis elegans: 4,493 proteins\n",
      "  Oryza sativa subsp. japonica (Rice): 4,195 proteins\n",
      "\n",
      "📏 Protein Length Statistics:\n",
      "  Min length: 2 amino acids\n",
      "  Max length: 35213 amino acids\n",
      "  Median length: 295 amino acids\n",
      "  Proteins >1000 AA: 19,269\n",
      "\n",
      " Domain Analysis:\n",
      "  Proteins with domains: 133,903 (23.3%)\n",
      "\n",
      " Coverage Summary:\n",
      "  High coverage (≥80%): 31 fields\n",
      "  Medium coverage (50-79%): 15 fields\n",
      "  Low coverage (10-49%): 36 fields\n",
      "  Very low coverage (<10%): 174 fields\n"
     ]
    }
   ],
   "source": [
    "def analyze_df(df):\n",
    "  \n",
    "  print(\"=== DATASET ANALYSIS ===\")\n",
    "  \n",
    "  # 1. Annotation coverage analysis\n",
    "  print(\"\\n Annotation Coverage:\")\n",
    "  high_coverage_fields = []\n",
    "  coverage_data = []\n",
    "  \n",
    "  for col in df.columns:\n",
    "      if col not in ['Entry', 'Entry Name', 'Sequence']:\n",
    "          coverage = df[col].notna().sum()\n",
    "          percentage = coverage / len(df) * 100\n",
    "          \n",
    "          coverage_data.append({\n",
    "              'field': col,\n",
    "              'coverage': coverage,\n",
    "              'percentage': percentage\n",
    "          })\n",
    "          \n",
    "          if percentage >= 50:\n",
    "              high_coverage_fields.append(col)\n",
    "  \n",
    "  # Sort by percentage (highest to lowest)\n",
    "  coverage_data.sort(key=lambda x: x['percentage'], reverse=True)\n",
    "  \n",
    "  # Print sorted coverage\n",
    "  for item in coverage_data:\n",
    "      col = item['field']\n",
    "      coverage = item['coverage']\n",
    "      percentage = item['percentage']\n",
    "      \n",
    "      # for visual purposes\n",
    "      if percentage >= 80:\n",
    "          indicator = \"🟢\"  \n",
    "      elif percentage >= 50:\n",
    "          indicator = \"🟡\"  \n",
    "      elif percentage >= 10:\n",
    "          indicator = \"🟠\"\n",
    "      else:\n",
    "          indicator = \"🔴\" \n",
    "          \n",
    "      print(f\"{indicator} {col}: {coverage:,}/{len(df):,} ({percentage:.1f}%)\")\n",
    "  \n",
    "  # 2. organsism diversity\n",
    "  print(f\"\\n Organism Diversity:\")\n",
    "  top_organisms = df['Organism'].value_counts().head(10)\n",
    "  print(\"Top 10 organisms:\")\n",
    "  for org, count in top_organisms.items():\n",
    "      print(f\"  {org}: {count:,} proteins\")\n",
    "  \n",
    "  # 3. protein length dist\n",
    "  print(f\"\\n📏 Protein Length Statistics:\")\n",
    "  print(f\"  Min length: {df['Length'].min()} amino acids\")\n",
    "  print(f\"  Max length: {df['Length'].max()} amino acids\")\n",
    "  print(f\"  Median length: {df['Length'].median():.0f} amino acids\")\n",
    "  print(f\"  Proteins >1000 AA: {(df['Length'] > 1000).sum():,}\")\n",
    "  \n",
    "  # 4. d0main analysis\n",
    "  if 'Domain [FT]' in df.columns:\n",
    "      domain_coverage = df['Domain [FT]'].notna().sum()\n",
    "      print(f\"\\n Domain Analysis:\")\n",
    "      print(f\"  Proteins with domains: {domain_coverage:,} ({domain_coverage/len(df)*100:.1f}%)\")\n",
    "  \n",
    "  # 5. Overall coverage summary\n",
    "  print(f\"\\n Coverage Summary:\")\n",
    "  high_cov = sum(1 for item in coverage_data if item['percentage'] >= 80)\n",
    "  medium_cov = sum(1 for item in coverage_data if 50 <= item['percentage'] < 80)\n",
    "  low_cov = sum(1 for item in coverage_data if 10 <= item['percentage'] < 50)\n",
    "  very_low_cov = sum(1 for item in coverage_data if item['percentage'] < 10)\n",
    "  \n",
    "  print(f\"  High coverage (≥80%): {high_cov} fields\")\n",
    "  print(f\"  Medium coverage (50-79%): {medium_cov} fields\") \n",
    "  print(f\"  Low coverage (10-49%): {low_cov} fields\")\n",
    "  print(f\"  Very low coverage (<10%): {very_low_cov} fields\")\n",
    "  \n",
    "  return {\n",
    "      'total_proteins': len(df),\n",
    "      'unique_sequences': df['Sequence'].nunique(),\n",
    "      'organisms': df['Organism'].nunique(),\n",
    "      'high_coverage_fields': high_coverage_fields,\n",
    "      'coverage_data': coverage_data \n",
    "  }\n",
    "\n",
    "analysis = analyze_df(df_full)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04aaaf3b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "myenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
