{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c758e028",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import requests\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7d92626e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_fasta(filepath):\n",
    "    records = []\n",
    "    with open(filepath, 'r') as f:\n",
    "        header = None\n",
    "        seq_lines = []\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):\n",
    "                if header:\n",
    "                    # Parse header fields\n",
    "                    parts = header.split(' ', 1)\n",
    "                    ids = parts[0].split('|')\n",
    "                    db = ids[0]\n",
    "                    accession = ids[1]\n",
    "                    entry_name = ids[2]\n",
    "                    rest = parts[1] if len(parts) > 1 else ''\n",
    "                    kv_pattern = r'(\\w+)=([^\\s]+)'\n",
    "                    kv_pairs = dict(re.findall(kv_pattern, rest))\n",
    "                    first_kv = re.search(kv_pattern, rest)\n",
    "                    protein_name = rest[:first_kv.start()].strip() if first_kv else rest.strip()\n",
    "                    records.append({\n",
    "                        'database': db,\n",
    "                        'accession_number': accession,\n",
    "                        'entry_name': entry_name,\n",
    "                        'protein_desc': protein_name,\n",
    "                        'organism_species': kv_pairs.get('OS', ''),\n",
    "                        'organism_identifier': kv_pairs.get('OX', ''),\n",
    "                        'gene_name': kv_pairs.get('GN', ''),\n",
    "                        'protein_existence_level': kv_pairs.get('PE', ''),\n",
    "                        'sequence_version': kv_pairs.get('SV', ''),\n",
    "                        'sequence': ''.join(seq_lines)\n",
    "                    })\n",
    "                header = line[1:]  # remove '>'\n",
    "                seq_lines = []\n",
    "            else:\n",
    "                seq_lines.append(line)\n",
    "        # Add the last record\n",
    "        if header:\n",
    "            parts = header.split(' ', 1)\n",
    "            ids = parts[0].split('|')\n",
    "            db = ids[0]\n",
    "            accession = ids[1]\n",
    "            entry_name = ids[2]\n",
    "            rest = parts[1] if len(parts) > 1 else ''\n",
    "            kv_pattern = r'(\\w+)=([^\\s]+)'\n",
    "            kv_pairs = dict(re.findall(kv_pattern, rest))\n",
    "            first_kv = re.search(kv_pattern, rest)\n",
    "            protein_name = rest[:first_kv.start()].strip() if first_kv else rest.strip()\n",
    "            records.append({\n",
    "                'database': db,\n",
    "                'accession_number': accession,\n",
    "                'entry_name': entry_name,\n",
    "                'protein_desc': protein_name,\n",
    "                'organism_species': kv_pairs.get('OS', ''),\n",
    "                'organism_identifier': kv_pairs.get('OX', ''),\n",
    "                'gene_name': kv_pairs.get('GN', ''),\n",
    "                'protein_existence_level': kv_pairs.get('PE', ''),\n",
    "                'sequence_version': kv_pairs.get('SV', ''),\n",
    "                'sequence': ''.join(seq_lines)\n",
    "            })\n",
    "    return pd.DataFrame(records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "252af6b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database</th>\n",
       "      <th>accession_number</th>\n",
       "      <th>protein_desc</th>\n",
       "      <th>organism_species</th>\n",
       "      <th>organism_identifier</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>protein_existence_level</th>\n",
       "      <th>sequence_version</th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>organism_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZX4</td>\n",
       "      <td>Putative transcription factor 001R</td>\n",
       "      <td>Frog</td>\n",
       "      <td>654924</td>\n",
       "      <td>FV3-001R</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>MAFSAEDVLKEYDRRRRMEALLLSLYYPNDRKLLDYKEWSPPRVQV...</td>\n",
       "      <td>001R</td>\n",
       "      <td>FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZX3</td>\n",
       "      <td>Uncharacterized protein 002L</td>\n",
       "      <td>Frog</td>\n",
       "      <td>654924</td>\n",
       "      <td>FV3-002L</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>MSIIGATRLQNDKSDTYSAGPCYAGGCSAFTPRGTCGKDWDLGEQT...</td>\n",
       "      <td>002L</td>\n",
       "      <td>FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp</td>\n",
       "      <td>Q197F8</td>\n",
       "      <td>Uncharacterized protein 002R</td>\n",
       "      <td>Invertebrate</td>\n",
       "      <td>345201</td>\n",
       "      <td>IIV3-002R</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>MASNTVSAQGGSNRPVRDFSNIQDVAQFLLFDPIWNEQPGSIVPWK...</td>\n",
       "      <td>002R</td>\n",
       "      <td>IIV3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp</td>\n",
       "      <td>Q197F7</td>\n",
       "      <td>Uncharacterized protein 003L</td>\n",
       "      <td>Invertebrate</td>\n",
       "      <td>345201</td>\n",
       "      <td>IIV3-003L</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>MYQAINPCPQSWYGSPQLEREIVCKMSGAPHYPNYYPVHPNALGGA...</td>\n",
       "      <td>003L</td>\n",
       "      <td>IIV3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp</td>\n",
       "      <td>Q6GZX2</td>\n",
       "      <td>Uncharacterized protein 3R</td>\n",
       "      <td>Frog</td>\n",
       "      <td>654924</td>\n",
       "      <td>FV3-003R</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...</td>\n",
       "      <td>003R</td>\n",
       "      <td>FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573656</th>\n",
       "      <td>sp</td>\n",
       "      <td>Q6UY62</td>\n",
       "      <td>RING finger protein Z</td>\n",
       "      <td>Sabia</td>\n",
       "      <td>3052299</td>\n",
       "      <td>Z</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGNSKSKSKLSANQYEQQTVNSTKQVAILKRQAEPSLYGRHNCRCC...</td>\n",
       "      <td>Z</td>\n",
       "      <td>SABVB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573657</th>\n",
       "      <td>sp</td>\n",
       "      <td>P08105</td>\n",
       "      <td>Putative uncharacterized protein Z</td>\n",
       "      <td>Ovis</td>\n",
       "      <td>9940</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>MSSSLEITSFYSFIWTPHIGPLLFGIGLWFSMFKEPSHFCPCQHPH...</td>\n",
       "      <td>Z</td>\n",
       "      <td>SHEEP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573658</th>\n",
       "      <td>sp</td>\n",
       "      <td>Q88470</td>\n",
       "      <td>RING finger protein Z</td>\n",
       "      <td>Tacaribe</td>\n",
       "      <td>928313</td>\n",
       "      <td>Z</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>MGNCNRTQKPSSSSNNLEKPPQAAEFRRTAEPSLYGRYNCKCCWFA...</td>\n",
       "      <td>Z</td>\n",
       "      <td>TACVF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573659</th>\n",
       "      <td>sp</td>\n",
       "      <td>A9JR22</td>\n",
       "      <td>RING finger protein Z</td>\n",
       "      <td>Tamiami</td>\n",
       "      <td>3052329</td>\n",
       "      <td>Z</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>MGLRYSKEVRDRHGDKDPEGRIPITQTMPQTLYGRYNCKSCWFANK...</td>\n",
       "      <td>Z</td>\n",
       "      <td>TAMVU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573660</th>\n",
       "      <td>sp</td>\n",
       "      <td>B2ZDY1</td>\n",
       "      <td>RING finger protein Z</td>\n",
       "      <td>Whitewater</td>\n",
       "      <td>3052331</td>\n",
       "      <td>Z</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>MGLRYSKDVKDRYGDREPEGRIPITLNMPQSLYGRYNCKSCWFANK...</td>\n",
       "      <td>Z</td>\n",
       "      <td>WWAVU</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>573661 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       database accession_number                        protein_desc  \\\n",
       "0            sp           Q6GZX4  Putative transcription factor 001R   \n",
       "1            sp           Q6GZX3        Uncharacterized protein 002L   \n",
       "2            sp           Q197F8        Uncharacterized protein 002R   \n",
       "3            sp           Q197F7        Uncharacterized protein 003L   \n",
       "4            sp           Q6GZX2          Uncharacterized protein 3R   \n",
       "...         ...              ...                                 ...   \n",
       "573656       sp           Q6UY62               RING finger protein Z   \n",
       "573657       sp           P08105  Putative uncharacterized protein Z   \n",
       "573658       sp           Q88470               RING finger protein Z   \n",
       "573659       sp           A9JR22               RING finger protein Z   \n",
       "573660       sp           B2ZDY1               RING finger protein Z   \n",
       "\n",
       "       organism_species organism_identifier  gene_name  \\\n",
       "0                  Frog              654924   FV3-001R   \n",
       "1                  Frog              654924   FV3-002L   \n",
       "2          Invertebrate              345201  IIV3-002R   \n",
       "3          Invertebrate              345201  IIV3-003L   \n",
       "4                  Frog              654924   FV3-003R   \n",
       "...                 ...                 ...        ...   \n",
       "573656            Sabia             3052299          Z   \n",
       "573657             Ovis                9940              \n",
       "573658         Tacaribe              928313          Z   \n",
       "573659          Tamiami             3052329          Z   \n",
       "573660       Whitewater             3052331          Z   \n",
       "\n",
       "       protein_existence_level sequence_version  \\\n",
       "0                            4                1   \n",
       "1                            4                1   \n",
       "2                            4                1   \n",
       "3                            4                1   \n",
       "4                            3                1   \n",
       "...                        ...              ...   \n",
       "573656                       1                1   \n",
       "573657                       4                1   \n",
       "573658                       1                3   \n",
       "573659                       3                1   \n",
       "573660                       3                1   \n",
       "\n",
       "                                                 sequence protein_id  \\\n",
       "0       MAFSAEDVLKEYDRRRRMEALLLSLYYPNDRKLLDYKEWSPPRVQV...       001R   \n",
       "1       MSIIGATRLQNDKSDTYSAGPCYAGGCSAFTPRGTCGKDWDLGEQT...       002L   \n",
       "2       MASNTVSAQGGSNRPVRDFSNIQDVAQFLLFDPIWNEQPGSIVPWK...       002R   \n",
       "3       MYQAINPCPQSWYGSPQLEREIVCKMSGAPHYPNYYPVHPNALGGA...       003L   \n",
       "4       MARPLLGKTSSVRRRLESLSACSIFFFLRKFCQKMASLVFLNSPVY...       003R   \n",
       "...                                                   ...        ...   \n",
       "573656  MGNSKSKSKLSANQYEQQTVNSTKQVAILKRQAEPSLYGRHNCRCC...          Z   \n",
       "573657  MSSSLEITSFYSFIWTPHIGPLLFGIGLWFSMFKEPSHFCPCQHPH...          Z   \n",
       "573658  MGNCNRTQKPSSSSNNLEKPPQAAEFRRTAEPSLYGRYNCKCCWFA...          Z   \n",
       "573659  MGLRYSKEVRDRHGDKDPEGRIPITQTMPQTLYGRYNCKSCWFANK...          Z   \n",
       "573660  MGLRYSKDVKDRYGDREPEGRIPITLNMPQSLYGRYNCKSCWFANK...          Z   \n",
       "\n",
       "       organism_id  \n",
       "0            FRG3G  \n",
       "1            FRG3G  \n",
       "2             IIV3  \n",
       "3             IIV3  \n",
       "4            FRG3G  \n",
       "...            ...  \n",
       "573656       SABVB  \n",
       "573657       SHEEP  \n",
       "573658       TACVF  \n",
       "573659       TAMVU  \n",
       "573660       WWAVU  \n",
       "\n",
       "[573661 rows x 11 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = parse_fasta(\"../data/swiss.fasta\")\n",
    "df[['protein_id', 'organism_id']] = df['entry_name'].str.split('_', n=1, expand=True)\n",
    "df = df.drop(columns=['entry_name'])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3eaa2991",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1it [00:01,  1.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'metadata': {'accession': 'IPR007031', 'name': 'Poxvirus VLTF3, late transcription factor', 'source_database': 'interpro', 'type': 'family', 'integrated': None, 'member_databases': {'pfam': {'PF04947': 'Poxvirus Late Transcription Factor VLTF3 like'}}, 'go_terms': [{'identifier': 'GO:0046782', 'name': 'regulation of viral transcription', 'category': {'code': 'P', 'name': 'biological_process'}}]}, 'proteins': [{'accession': 'q6gzx4', 'protein_length': 256, 'source_database': 'reviewed', 'organism': '654924', 'in_alphafold': False, 'in_bfvd': False, 'entry_protein_locations': [{'fragments': [{'start': 81, 'end': 253, 'dc-status': 'CONTINUOUS'}], 'representative': False, 'model': None, 'score': None}]}]}]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2it [00:01,  1.27it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'metadata': {'accession': 'IPR004251', 'name': 'Pox virus entry-fusion-complex G9/A16', 'source_database': 'interpro', 'type': 'family', 'integrated': None, 'member_databases': {'pfam': {'PF03003': 'Pox virus entry-fusion-complex G9/A16'}}, 'go_terms': None}, 'proteins': [{'accession': 'q6gzx3', 'protein_length': 320, 'source_database': 'reviewed', 'organism': '654924', 'in_alphafold': False, 'in_bfvd': False, 'entry_protein_locations': [{'fragments': [{'start': 169, 'end': 247, 'dc-status': 'CONTINUOUS'}], 'representative': False, 'model': None, 'score': None}]}]}]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "8it [00:05,  1.57it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'metadata': {'accession': 'IPR003360', 'name': 'US22-like', 'source_database': 'interpro', 'type': 'family', 'integrated': None, 'member_databases': {'pfam': {'PF02393': 'US22 like'}}, 'go_terms': None}, 'proteins': [{'accession': 'q6gzx0', 'protein_length': 204, 'source_database': 'reviewed', 'organism': '654924', 'in_alphafold': False, 'in_bfvd': False, 'entry_protein_locations': [{'fragments': [{'start': 37, 'end': 150, 'dc-status': 'CONTINUOUS'}], 'representative': False, 'model': None, 'score': None}]}]}]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "9it [00:06,  1.58it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'metadata': {'accession': 'IPR017880', 'name': 'KilA, N-terminal', 'source_database': 'interpro', 'type': 'domain', 'integrated': None, 'member_databases': {'profile': {'PS51301': 'KilA-N domain profile'}}, 'go_terms': None}, 'proteins': [{'accession': 'q91g88', 'protein_length': 352, 'source_database': 'reviewed', 'organism': '176652', 'in_alphafold': False, 'in_bfvd': False, 'entry_protein_locations': [{'fragments': [{'start': 15, 'end': 123, 'dc-status': 'CONTINUOUS'}], 'representative': False, 'model': None, 'score': None}]}]}, {'metadata': {'accession': 'IPR018004', 'name': 'KilA/APSES-type HTH, DNA-binding', 'source_database': 'interpro', 'type': 'domain', 'integrated': None, 'member_databases': {'smart': {'SM01252': 'KilA-N'}, 'pfam': {'PF04383': 'KilA-N domain'}}, 'go_terms': None}, 'proteins': [{'accession': 'q91g88', 'protein_length': 352, 'source_database': 'reviewed', 'organism': '176652', 'in_alphafold': False, 'in_bfvd': False, 'entry_protein_locations': [{'fragments': [{'start': 21, 'end': 123, 'dc-status': 'CONTINUOUS'}], 'representative': False, 'model': None, 'score': None}]}]}, {'metadata': {'accession': 'IPR022549', 'name': 'Domain of unknown function DUF3627', 'source_database': 'interpro', 'type': 'domain', 'integrated': None, 'member_databases': {'pfam': {'PF12299': 'Protein of unknown function (DUF3627)'}}, 'go_terms': None}, 'proteins': [{'accession': 'q91g88', 'protein_length': 352, 'source_database': 'reviewed', 'organism': '176652', 'in_alphafold': False, 'in_bfvd': False, 'entry_protein_locations': [{'fragments': [{'start': 231, 'end': 322, 'dc-status': 'CONTINUOUS'}], 'representative': False, 'model': None, 'score': None}]}]}]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "9it [00:06,  1.35it/s]\n"
     ]
    }
   ],
   "source": [
    "def fetch_interpro_annotations(accession):\n",
    "    url = f\"https://www.ebi.ac.uk/interpro/api/entry/interpro/protein/uniprot/{accession}/\"\n",
    "    headers = {\"Accept\": \"application/json\"}\n",
    "    response = requests.get(url, headers=headers)\n",
    "\n",
    "    if response.status_code != 200:\n",
    "        return []\n",
    "\n",
    "    data = response.json().get(\"results\", [])\n",
    "    print(data)\n",
    "    annotations = []\n",
    "    for entry in data:\n",
    "        meta = entry.get(\"metadata\", {})\n",
    "        interpro_id = meta.get(\"accession\", \"\")\n",
    "        interpro_name = meta.get(\"name\", \"\")\n",
    "        source_database = meta.get(\"source_database\", \"\")\n",
    "        interpro_type = meta.get(\"type\", \"\")\n",
    "        integrated = meta.get(\"integrated\", None)\n",
    "        member_databases = meta.get(\"member_databases\", {})\n",
    "        go_terms = meta.get(\"go_terms\", None)\n",
    "\n",
    "        for protein in entry.get(\"proteins\", []):\n",
    "            for loc in protein.get(\"entry_protein_locations\", []):\n",
    "                for fragment in loc.get(\"fragments\", []):\n",
    "                    annotations.append({\n",
    "                        \"interpro_id\": interpro_id,\n",
    "                        \"interpro_name\": interpro_name,\n",
    "                        \"source_database\": source_database,\n",
    "                        \"interpro_type\": interpro_type,\n",
    "                        \"integrated\": integrated,\n",
    "                        \"member_databases\": member_databases,\n",
    "                        \"go_terms\": go_terms,\n",
    "                        \"start\": fragment.get(\"start\"),\n",
    "                        \"end\": fragment.get(\"end\")\n",
    "                    })\n",
    "    return annotations\n",
    "\n",
    "def get_interpro_dataframe_from_fasta(df):\n",
    "    expanded_rows = []\n",
    "    i = 0\n",
    "    for _, row in tqdm(df.iterrows()):\n",
    "        accession = row['accession_number']\n",
    "        annotations = fetch_interpro_annotations(accession)\n",
    "        for ann in annotations:\n",
    "            expanded_rows.append({\n",
    "                \"accession_number\": accession,\n",
    "                \"protein_id\": row['protein_id'],\n",
    "                \"organism_id\": row['organism_id'],\n",
    "                **ann\n",
    "            })\n",
    "        i += 1\n",
    "        if i == 10:\n",
    "            break\n",
    "\n",
    "    return pd.DataFrame(expanded_rows)\n",
    "\n",
    "enriched_df = get_interpro_dataframe_from_fasta(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7fe0c3ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_number</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>organism_id</th>\n",
       "      <th>interpro_id</th>\n",
       "      <th>interpro_name</th>\n",
       "      <th>source_database</th>\n",
       "      <th>interpro_type</th>\n",
       "      <th>integrated</th>\n",
       "      <th>member_databases</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q6GZX4</td>\n",
       "      <td>001R</td>\n",
       "      <td>FRG3G</td>\n",
       "      <td>IPR007031</td>\n",
       "      <td>Poxvirus VLTF3, late transcription factor</td>\n",
       "      <td>interpro</td>\n",
       "      <td>family</td>\n",
       "      <td>None</td>\n",
       "      <td>{'pfam': {'PF04947': 'Poxvirus Late Transcript...</td>\n",
       "      <td>[{'identifier': 'GO:0046782', 'name': 'regulat...</td>\n",
       "      <td>81</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q6GZX3</td>\n",
       "      <td>002L</td>\n",
       "      <td>FRG3G</td>\n",
       "      <td>IPR004251</td>\n",
       "      <td>Pox virus entry-fusion-complex G9/A16</td>\n",
       "      <td>interpro</td>\n",
       "      <td>family</td>\n",
       "      <td>None</td>\n",
       "      <td>{'pfam': {'PF03003': 'Pox virus entry-fusion-c...</td>\n",
       "      <td>None</td>\n",
       "      <td>169</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q6GZX0</td>\n",
       "      <td>005R</td>\n",
       "      <td>FRG3G</td>\n",
       "      <td>IPR003360</td>\n",
       "      <td>US22-like</td>\n",
       "      <td>interpro</td>\n",
       "      <td>family</td>\n",
       "      <td>None</td>\n",
       "      <td>{'pfam': {'PF02393': 'US22 like'}}</td>\n",
       "      <td>None</td>\n",
       "      <td>37</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q91G88</td>\n",
       "      <td>006L</td>\n",
       "      <td>IIV6</td>\n",
       "      <td>IPR017880</td>\n",
       "      <td>KilA, N-terminal</td>\n",
       "      <td>interpro</td>\n",
       "      <td>domain</td>\n",
       "      <td>None</td>\n",
       "      <td>{'profile': {'PS51301': 'KilA-N domain profile'}}</td>\n",
       "      <td>None</td>\n",
       "      <td>15</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q91G88</td>\n",
       "      <td>006L</td>\n",
       "      <td>IIV6</td>\n",
       "      <td>IPR018004</td>\n",
       "      <td>KilA/APSES-type HTH, DNA-binding</td>\n",
       "      <td>interpro</td>\n",
       "      <td>domain</td>\n",
       "      <td>None</td>\n",
       "      <td>{'smart': {'SM01252': 'KilA-N'}, 'pfam': {'PF0...</td>\n",
       "      <td>None</td>\n",
       "      <td>21</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Q91G88</td>\n",
       "      <td>006L</td>\n",
       "      <td>IIV6</td>\n",
       "      <td>IPR022549</td>\n",
       "      <td>Domain of unknown function DUF3627</td>\n",
       "      <td>interpro</td>\n",
       "      <td>domain</td>\n",
       "      <td>None</td>\n",
       "      <td>{'pfam': {'PF12299': 'Protein of unknown funct...</td>\n",
       "      <td>None</td>\n",
       "      <td>231</td>\n",
       "      <td>322</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  accession_number protein_id organism_id interpro_id  \\\n",
       "0           Q6GZX4       001R       FRG3G   IPR007031   \n",
       "1           Q6GZX3       002L       FRG3G   IPR004251   \n",
       "2           Q6GZX0       005R       FRG3G   IPR003360   \n",
       "3           Q91G88       006L        IIV6   IPR017880   \n",
       "4           Q91G88       006L        IIV6   IPR018004   \n",
       "5           Q91G88       006L        IIV6   IPR022549   \n",
       "\n",
       "                               interpro_name source_database interpro_type  \\\n",
       "0  Poxvirus VLTF3, late transcription factor        interpro        family   \n",
       "1      Pox virus entry-fusion-complex G9/A16        interpro        family   \n",
       "2                                  US22-like        interpro        family   \n",
       "3                           KilA, N-terminal        interpro        domain   \n",
       "4           KilA/APSES-type HTH, DNA-binding        interpro        domain   \n",
       "5         Domain of unknown function DUF3627        interpro        domain   \n",
       "\n",
       "  integrated                                   member_databases  \\\n",
       "0       None  {'pfam': {'PF04947': 'Poxvirus Late Transcript...   \n",
       "1       None  {'pfam': {'PF03003': 'Pox virus entry-fusion-c...   \n",
       "2       None                 {'pfam': {'PF02393': 'US22 like'}}   \n",
       "3       None  {'profile': {'PS51301': 'KilA-N domain profile'}}   \n",
       "4       None  {'smart': {'SM01252': 'KilA-N'}, 'pfam': {'PF0...   \n",
       "5       None  {'pfam': {'PF12299': 'Protein of unknown funct...   \n",
       "\n",
       "                                            go_terms  start  end  \n",
       "0  [{'identifier': 'GO:0046782', 'name': 'regulat...     81  253  \n",
       "1                                               None    169  247  \n",
       "2                                               None     37  150  \n",
       "3                                               None     15  123  \n",
       "4                                               None     21  123  \n",
       "5                                               None    231  322  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enriched_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f4c3110d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge dataframes\n",
    "merged_df = pd.merge(\n",
    "    enriched_df,\n",
    "    df,\n",
    "    on=['accession_number'],\n",
    "    how='left'  # or 'inner' if you only want rows with matches in both\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "80a755a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_number</th>\n",
       "      <th>protein_id_x</th>\n",
       "      <th>organism_id_x</th>\n",
       "      <th>interpro_id</th>\n",
       "      <th>interpro_name</th>\n",
       "      <th>source_database</th>\n",
       "      <th>interpro_type</th>\n",
       "      <th>integrated</th>\n",
       "      <th>member_databases</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>...</th>\n",
       "      <th>database</th>\n",
       "      <th>protein_desc</th>\n",
       "      <th>organism_species</th>\n",
       "      <th>organism_identifier</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>protein_existence_level</th>\n",
       "      <th>sequence_version</th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein_id_y</th>\n",
       "      <th>organism_id_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q6GZX4</td>\n",
       "      <td>001R</td>\n",
       "      <td>FRG3G</td>\n",
       "      <td>IPR007031</td>\n",
       "      <td>Poxvirus VLTF3, late transcription factor</td>\n",
       "      <td>interpro</td>\n",
       "      <td>family</td>\n",
       "      <td>None</td>\n",
       "      <td>{'pfam': {'PF04947': 'Poxvirus Late Transcript...</td>\n",
       "      <td>[{'identifier': 'GO:0046782', 'name': 'regulat...</td>\n",
       "      <td>...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Putative transcription factor 001R</td>\n",
       "      <td>Frog</td>\n",
       "      <td>654924</td>\n",
       "      <td>FV3-001R</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>MAFSAEDVLKEYDRRRRMEALLLSLYYPNDRKLLDYKEWSPPRVQV...</td>\n",
       "      <td>001R</td>\n",
       "      <td>FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q6GZX3</td>\n",
       "      <td>002L</td>\n",
       "      <td>FRG3G</td>\n",
       "      <td>IPR004251</td>\n",
       "      <td>Pox virus entry-fusion-complex G9/A16</td>\n",
       "      <td>interpro</td>\n",
       "      <td>family</td>\n",
       "      <td>None</td>\n",
       "      <td>{'pfam': {'PF03003': 'Pox virus entry-fusion-c...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Uncharacterized protein 002L</td>\n",
       "      <td>Frog</td>\n",
       "      <td>654924</td>\n",
       "      <td>FV3-002L</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>MSIIGATRLQNDKSDTYSAGPCYAGGCSAFTPRGTCGKDWDLGEQT...</td>\n",
       "      <td>002L</td>\n",
       "      <td>FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q6GZX0</td>\n",
       "      <td>005R</td>\n",
       "      <td>FRG3G</td>\n",
       "      <td>IPR003360</td>\n",
       "      <td>US22-like</td>\n",
       "      <td>interpro</td>\n",
       "      <td>family</td>\n",
       "      <td>None</td>\n",
       "      <td>{'pfam': {'PF02393': 'US22 like'}}</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Uncharacterized protein 005R</td>\n",
       "      <td>Frog</td>\n",
       "      <td>654924</td>\n",
       "      <td>FV3-005R</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>MQNPLPEVMSPEHDKRTTTPMSKEANKFIRELDKKPGDLAVVSDFV...</td>\n",
       "      <td>005R</td>\n",
       "      <td>FRG3G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q91G88</td>\n",
       "      <td>006L</td>\n",
       "      <td>IIV6</td>\n",
       "      <td>IPR017880</td>\n",
       "      <td>KilA, N-terminal</td>\n",
       "      <td>interpro</td>\n",
       "      <td>domain</td>\n",
       "      <td>None</td>\n",
       "      <td>{'profile': {'PS51301': 'KilA-N domain profile'}}</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Putative KilA-N domain-containing protein 006L</td>\n",
       "      <td>Invertebrate</td>\n",
       "      <td>176652</td>\n",
       "      <td>IIV6-006L</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>MDSLNEVCYEQIKGTFYKGLFGDFPLIVDKKTGCFNATKLCVLGGK...</td>\n",
       "      <td>006L</td>\n",
       "      <td>IIV6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Q91G88</td>\n",
       "      <td>006L</td>\n",
       "      <td>IIV6</td>\n",
       "      <td>IPR018004</td>\n",
       "      <td>KilA/APSES-type HTH, DNA-binding</td>\n",
       "      <td>interpro</td>\n",
       "      <td>domain</td>\n",
       "      <td>None</td>\n",
       "      <td>{'smart': {'SM01252': 'KilA-N'}, 'pfam': {'PF0...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Putative KilA-N domain-containing protein 006L</td>\n",
       "      <td>Invertebrate</td>\n",
       "      <td>176652</td>\n",
       "      <td>IIV6-006L</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>MDSLNEVCYEQIKGTFYKGLFGDFPLIVDKKTGCFNATKLCVLGGK...</td>\n",
       "      <td>006L</td>\n",
       "      <td>IIV6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Q91G88</td>\n",
       "      <td>006L</td>\n",
       "      <td>IIV6</td>\n",
       "      <td>IPR022549</td>\n",
       "      <td>Domain of unknown function DUF3627</td>\n",
       "      <td>interpro</td>\n",
       "      <td>domain</td>\n",
       "      <td>None</td>\n",
       "      <td>{'pfam': {'PF12299': 'Protein of unknown funct...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>sp</td>\n",
       "      <td>Putative KilA-N domain-containing protein 006L</td>\n",
       "      <td>Invertebrate</td>\n",
       "      <td>176652</td>\n",
       "      <td>IIV6-006L</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>MDSLNEVCYEQIKGTFYKGLFGDFPLIVDKKTGCFNATKLCVLGGK...</td>\n",
       "      <td>006L</td>\n",
       "      <td>IIV6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  accession_number protein_id_x organism_id_x interpro_id  \\\n",
       "0           Q6GZX4         001R         FRG3G   IPR007031   \n",
       "1           Q6GZX3         002L         FRG3G   IPR004251   \n",
       "2           Q6GZX0         005R         FRG3G   IPR003360   \n",
       "3           Q91G88         006L          IIV6   IPR017880   \n",
       "4           Q91G88         006L          IIV6   IPR018004   \n",
       "5           Q91G88         006L          IIV6   IPR022549   \n",
       "\n",
       "                               interpro_name source_database interpro_type  \\\n",
       "0  Poxvirus VLTF3, late transcription factor        interpro        family   \n",
       "1      Pox virus entry-fusion-complex G9/A16        interpro        family   \n",
       "2                                  US22-like        interpro        family   \n",
       "3                           KilA, N-terminal        interpro        domain   \n",
       "4           KilA/APSES-type HTH, DNA-binding        interpro        domain   \n",
       "5         Domain of unknown function DUF3627        interpro        domain   \n",
       "\n",
       "  integrated                                   member_databases  \\\n",
       "0       None  {'pfam': {'PF04947': 'Poxvirus Late Transcript...   \n",
       "1       None  {'pfam': {'PF03003': 'Pox virus entry-fusion-c...   \n",
       "2       None                 {'pfam': {'PF02393': 'US22 like'}}   \n",
       "3       None  {'profile': {'PS51301': 'KilA-N domain profile'}}   \n",
       "4       None  {'smart': {'SM01252': 'KilA-N'}, 'pfam': {'PF0...   \n",
       "5       None  {'pfam': {'PF12299': 'Protein of unknown funct...   \n",
       "\n",
       "                                            go_terms  ...  database  \\\n",
       "0  [{'identifier': 'GO:0046782', 'name': 'regulat...  ...        sp   \n",
       "1                                               None  ...        sp   \n",
       "2                                               None  ...        sp   \n",
       "3                                               None  ...        sp   \n",
       "4                                               None  ...        sp   \n",
       "5                                               None  ...        sp   \n",
       "\n",
       "                                     protein_desc organism_species  \\\n",
       "0              Putative transcription factor 001R             Frog   \n",
       "1                    Uncharacterized protein 002L             Frog   \n",
       "2                    Uncharacterized protein 005R             Frog   \n",
       "3  Putative KilA-N domain-containing protein 006L     Invertebrate   \n",
       "4  Putative KilA-N domain-containing protein 006L     Invertebrate   \n",
       "5  Putative KilA-N domain-containing protein 006L     Invertebrate   \n",
       "\n",
       "  organism_identifier  gene_name protein_existence_level sequence_version  \\\n",
       "0              654924   FV3-001R                       4                1   \n",
       "1              654924   FV3-002L                       4                1   \n",
       "2              654924   FV3-005R                       4                1   \n",
       "3              176652  IIV6-006L                       3                1   \n",
       "4              176652  IIV6-006L                       3                1   \n",
       "5              176652  IIV6-006L                       3                1   \n",
       "\n",
       "                                            sequence protein_id_y  \\\n",
       "0  MAFSAEDVLKEYDRRRRMEALLLSLYYPNDRKLLDYKEWSPPRVQV...         001R   \n",
       "1  MSIIGATRLQNDKSDTYSAGPCYAGGCSAFTPRGTCGKDWDLGEQT...         002L   \n",
       "2  MQNPLPEVMSPEHDKRTTTPMSKEANKFIRELDKKPGDLAVVSDFV...         005R   \n",
       "3  MDSLNEVCYEQIKGTFYKGLFGDFPLIVDKKTGCFNATKLCVLGGK...         006L   \n",
       "4  MDSLNEVCYEQIKGTFYKGLFGDFPLIVDKKTGCFNATKLCVLGGK...         006L   \n",
       "5  MDSLNEVCYEQIKGTFYKGLFGDFPLIVDKKTGCFNATKLCVLGGK...         006L   \n",
       "\n",
       "  organism_id_y  \n",
       "0         FRG3G  \n",
       "1         FRG3G  \n",
       "2         FRG3G  \n",
       "3          IIV6  \n",
       "4          IIV6  \n",
       "5          IIV6  \n",
       "\n",
       "[6 rows x 22 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "40515f5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mWARNING: Ignoring invalid distribution -cikit-learn (/Users/davidm/opt/anaconda3/lib/python3.9/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0mRequirement already satisfied: modlamp in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (4.3.2)\n",
      "Requirement already satisfied: setuptools>=20.2.2 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (75.8.0)\n",
      "Requirement already satisfied: nose>=1.3.7 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.3.7)\n",
      "Requirement already satisfied: numpy>=1.10.4 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.24.4)\n",
      "Requirement already satisfied: scipy>=0.17.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.9.1)\n",
      "Requirement already satisfied: matplotlib>=1.5.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (3.5.2)\n",
      "Requirement already satisfied: scikit-learn>=0.18.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.4.2)\n",
      "Requirement already satisfied: pandas>=0.18.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.3.5)\n",
      "Requirement already satisfied: requests>=2.11.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (2.32.3)\n",
      "Requirement already satisfied: lxml>=3.6.4 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (4.9.1)\n",
      "Requirement already satisfied: joblib>=0.15.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.4.2)\n",
      "Requirement already satisfied: mysql-connector-python>=8.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (9.4.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (0.11.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (4.25.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (1.4.2)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (21.3)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (9.2.0)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (3.2.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2017.3 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from pandas>=0.18.1->modlamp) (2022.1)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (3.3)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (1.26.11)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (2022.9.24)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from scikit-learn>=0.18.0->modlamp) (3.5.0)\n",
      "Requirement already satisfied: six>=1.5 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from python-dateutil>=2.7->matplotlib>=1.5.1->modlamp) (1.16.0)\n",
      "\u001b[33mWARNING: Ignoring invalid distribution -cikit-learn (/Users/davidm/opt/anaconda3/lib/python3.9/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[33mWARNING: Ignoring invalid distribution -cikit-learn (/Users/davidm/opt/anaconda3/lib/python3.9/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m25.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.1.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install modlamp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "727cf0d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mWARNING: Ignoring invalid distribution -cikit-learn (/Users/davidm/opt/anaconda3/lib/python3.9/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0mRequirement already satisfied: modlamp in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (4.3.2)\n",
      "Requirement already satisfied: setuptools>=20.2.2 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (75.8.0)\n",
      "Requirement already satisfied: nose>=1.3.7 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.3.7)\n",
      "Requirement already satisfied: numpy>=1.10.4 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.24.4)\n",
      "Requirement already satisfied: scipy>=0.17.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.9.1)\n",
      "Requirement already satisfied: matplotlib>=1.5.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (3.5.2)\n",
      "Requirement already satisfied: scikit-learn>=0.18.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.4.2)\n",
      "Requirement already satisfied: pandas>=0.18.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.3.5)\n",
      "Requirement already satisfied: requests>=2.11.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (2.32.3)\n",
      "Requirement already satisfied: lxml>=3.6.4 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (4.9.1)\n",
      "Requirement already satisfied: joblib>=0.15.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (1.4.2)\n",
      "Requirement already satisfied: mysql-connector-python>=8.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from modlamp) (9.4.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (0.11.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (4.25.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (1.4.2)\n",
      "Requirement already satisfied: packaging>=20.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (21.3)\n",
      "Requirement already satisfied: pillow>=6.2.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (9.2.0)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (3.2.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5.1->modlamp) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2017.3 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from pandas>=0.18.1->modlamp) (2022.1)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (3.3)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (1.26.11)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from requests>=2.11.1->modlamp) (2022.9.24)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from scikit-learn>=0.18.0->modlamp) (3.5.0)\n",
      "Requirement already satisfied: six>=1.5 in /Users/davidm/opt/anaconda3/lib/python3.9/site-packages (from python-dateutil>=2.7->matplotlib>=1.5.1->modlamp) (1.16.0)\n",
      "\u001b[33mWARNING: Ignoring invalid distribution -cikit-learn (/Users/davidm/opt/anaconda3/lib/python3.9/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0m\u001b[33mWARNING: Ignoring invalid distribution -cikit-learn (/Users/davidm/opt/anaconda3/lib/python3.9/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m25.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.1.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "! pip install modlamp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "298177fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Optional, Dict\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "\n",
    "\n",
    "# Fix ambiguous amino acids\n",
    "AA_FIXES = {\n",
    "    'B': 'D', 'Z': 'E', 'J': 'L',\n",
    "    'U': 'C', 'O': 'K', 'X': 'G',\n",
    "}\n",
    "VALID_AAS = set(\"ACDEFGHIKLMNPQRSTVWY\")\n",
    "\n",
    "\n",
    "def clean_sequence(seq: str) -> str:\n",
    "    \"\"\"Replace ambiguous amino acids with nearest valid equivalents.\"\"\"\n",
    "    return ''.join(AA_FIXES.get(aa, aa) for aa in seq if aa in VALID_AAS or aa in AA_FIXES)\n",
    "\n",
    "\n",
    "def compute_biopython_features(seq: str, verbose: bool = False) -> Optional[Dict[str, float]]:\n",
    "    \"\"\"\n",
    "    Compute amino acid features using Biopython and optionally modlamp.\n",
    "\n",
    "    Args:\n",
    "        seq (str): Original amino acid sequence.\n",
    "        verbose (bool): If True, print error messages.\n",
    "\n",
    "    Returns:\n",
    "        dict or None: Feature dictionary or None if failed.\n",
    "    \"\"\"\n",
    "    cleaned_seq = clean_sequence(seq)\n",
    "    \n",
    "    if len(cleaned_seq) < 3:\n",
    "        if verbose:\n",
    "            print(f\"[Too short after cleaning] {seq}\")\n",
    "        return None\n",
    "\n",
    "    try:\n",
    "        analysis = ProteinAnalysis(cleaned_seq)\n",
    "        aa_counts = analysis.count_amino_acids()\n",
    "        aa_percents = analysis.amino_acids_percent\n",
    "        sec_frac = analysis.secondary_structure_fraction()\n",
    "\n",
    "\n",
    "        features = {\n",
    "            \"Sequence\": seq,\n",
    "            \"Cleaned_Sequence\": cleaned_seq,\n",
    "            \"length\": len(cleaned_seq),\n",
    "            \"mol_weight\": analysis.molecular_weight(),\n",
    "            \"iso_point\": analysis.isoelectric_point(),\n",
    "            \"aromaticity\": analysis.aromaticity(),\n",
    "            \"instability_index\": analysis.instability_index(),\n",
    "            \"gravy\": analysis.gravy(),\n",
    "            \"helix_frac\": sec_frac[0],\n",
    "            \"turn_frac\": sec_frac[1],\n",
    "            \"sheet_frac\": sec_frac[2],\n",
    "        }\n",
    "\n",
    "        for aa in aa_counts:\n",
    "            features[f\"count_{aa}\"] = aa_counts[aa]\n",
    "            features[f\"percent_{aa}\"] = round(aa_percents[aa], 3)\n",
    "\n",
    "    except Exception as e:\n",
    "        if verbose:\n",
    "            print(f\"[Biopython Error] {seq}: {e}\")\n",
    "        return None\n",
    "\n",
    "    return features\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a61fbcdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Sequence': 'MAFSAEDVLKEYDRRRRMEALLLSLYYPNDRKLLDYKEWSPPRVQVECPKAPVEWNNPPSEKGLIVGHFSGIKYKGEKAQASEVDVNKMCCWVSKFKDAMRRYQGIQTCKIPGKVLSDLDAKIKAYNLTVEGVEGFVRYSRVTKQHVAAFLKELRHSKQYENVNLIHYILTDKRVDIQHLEKDLVKDFKALVESAHRMRQGHMINVKYILYQLLKKHGHGPDGPDILTVKTGSKGVLYDDSFRKIYTDLGWKFTPL',\n",
       " 'Cleaned_Sequence': 'MAFSAEDVLKEYDRRRRMEALLLSLYYPNDRKLLDYKEWSPPRVQVECPKAPVEWNNPPSEKGLIVGHFSGIKYKGEKAQASEVDVNKMCCWVSKFKDAMRRYQGIQTCKIPGKVLSDLDAKIKAYNLTVEGVEGFVRYSRVTKQHVAAFLKELRHSKQYENVNLIHYILTDKRVDIQHLEKDLVKDFKALVESAHRMRQGHMINVKYILYQLLKKHGHGPDGPDILTVKTGSKGVLYDDSFRKIYTDLGWKFTPL',\n",
       " 'length': 256,\n",
       " 'mol_weight': 29735.10070000003,\n",
       " 'iso_point': 9.370173454284668,\n",
       " 'aromaticity': 0.1015625,\n",
       " 'instability_index': 31.205078124999996,\n",
       " 'gravy': -0.5386718750000006,\n",
       " 'helix_frac': 0.34375,\n",
       " 'turn_frac': 0.25,\n",
       " 'sheet_frac': 0.359375,\n",
       " 'count_A': 13,\n",
       " 'percent_A': 5.078,\n",
       " 'count_C': 4,\n",
       " 'percent_C': 1.562,\n",
       " 'count_D': 17,\n",
       " 'percent_D': 6.641,\n",
       " 'count_E': 15,\n",
       " 'percent_E': 5.859,\n",
       " 'count_F': 8,\n",
       " 'percent_F': 3.125,\n",
       " 'count_G': 15,\n",
       " 'percent_G': 5.859,\n",
       " 'count_H': 9,\n",
       " 'percent_H': 3.516,\n",
       " 'count_I': 12,\n",
       " 'percent_I': 4.688,\n",
       " 'count_K': 29,\n",
       " 'percent_K': 11.328,\n",
       " 'count_L': 25,\n",
       " 'percent_L': 9.766,\n",
       " 'count_M': 6,\n",
       " 'percent_M': 2.344,\n",
       " 'count_N': 8,\n",
       " 'percent_N': 3.125,\n",
       " 'count_P': 11,\n",
       " 'percent_P': 4.297,\n",
       " 'count_Q': 9,\n",
       " 'percent_Q': 3.516,\n",
       " 'count_R': 15,\n",
       " 'percent_R': 5.859,\n",
       " 'count_S': 13,\n",
       " 'percent_S': 5.078,\n",
       " 'count_T': 8,\n",
       " 'percent_T': 3.125,\n",
       " 'count_V': 21,\n",
       " 'percent_V': 8.203,\n",
       " 'count_W': 4,\n",
       " 'percent_W': 1.562,\n",
       " 'count_Y': 14,\n",
       " 'percent_Y': 5.469}"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bio_features = merged_df[\"sequence\"].apply(compute_biopython_features)\n",
    "bio_features[0] # sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "c6be7dc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from logging import getLogger\n",
    "from typing import Dict, List, Optional, Set, Tuple\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "logger = getLogger(__name__)\n",
    "\n",
    "\n",
    "def process_binary_feature(\n",
    "    column_data: str, column_name: str, seq_len: int, current_index: int\n",
    ") -> Tuple[List[bool], int]:\n",
    "    \"\"\"\n",
    "    Process binary features from UniProt annotations (i.e. Helix, Turn)\n",
    "    \"\"\"\n",
    "    indices = [False] * seq_len\n",
    "    if pd.isna(column_data):\n",
    "        return indices, current_index\n",
    "    entries = column_data.split(f\"{column_name} \")[1:]\n",
    "    for entry in entries:\n",
    "        index = entry.split(\";\")[0]\n",
    "        try:\n",
    "            if \":\" in index or \"?\" in index:\n",
    "                continue\n",
    "            if \"..\" in index:\n",
    "                start, end = index.split(\"..\")\n",
    "                start = start.strip(\"<\")\n",
    "                end = end.strip(\">\")\n",
    "                start, end = int(start), int(end)\n",
    "                for i in range(start - 1, min(end, seq_len)):\n",
    "                    indices[i] = current_index\n",
    "            else:\n",
    "                idx = int(index) - 1\n",
    "                if 0 <= idx < seq_len:\n",
    "                    indices[idx] = current_index\n",
    "\n",
    "            current_index += 1\n",
    "        except Exception as e:\n",
    "            print(f\"Error processing binary column {column_name}: {e}\")\n",
    "            print(f\"Column data: {column_data}, Index: {index}\")\n",
    "    return indices, current_index\n",
    "\n",
    "\n",
    "def process_interaction_feature(\n",
    "    column_data: str, column_name: str, seq_len: int\n",
    ") -> Tuple[List[bool], List[Optional[int]]]:\n",
    "    \"\"\"\n",
    "    Process interaction features (i.e. disulfide bonds) from UniProt annotations.\n",
    "    \"\"\"\n",
    "    indices = [False] * seq_len\n",
    "    pairs = [None] * seq_len\n",
    "    if pd.isna(column_data):\n",
    "        return indices, pairs\n",
    "    n_pairs = 0\n",
    "    entries = column_data.split(f\"{column_name} \")[1:]\n",
    "    for entry in entries:\n",
    "        index = entry.split(\";\")[0]\n",
    "        try:\n",
    "            if \":\" in index or \"?\" in index:\n",
    "                continue\n",
    "            if \"..\" in index:\n",
    "                start, end = index.split(\"..\")\n",
    "                start = int(start.strip(\"<\"))\n",
    "                end = int(end.strip(\">\"))\n",
    "\n",
    "                if 0 <= start - 1 < seq_len:\n",
    "                    indices[start - 1] = True\n",
    "                    pairs[start - 1] = n_pairs\n",
    "                if 0 <= end - 1 < seq_len:\n",
    "                    indices[end - 1] = True\n",
    "                    pairs[end - 1] = n_pairs\n",
    "                n_pairs += 1\n",
    "            else:\n",
    "                idx = int(index) - 1\n",
    "                if 0 <= idx < seq_len:\n",
    "                    indices[idx] = True\n",
    "        except Exception as e:\n",
    "            print(f\"Error processing interaction column {column_name}: {e}\")\n",
    "            print(f\"Column data: {column_data}, Index: {index}\")\n",
    "    return indices, pairs\n",
    "\n",
    "\n",
    "def process_categorical_feature(\n",
    "    column_data: str,\n",
    "    column_name: str,\n",
    "    category_options: Set[str],\n",
    "    seq_len: int,\n",
    "    current_index: Dict[str, int],\n",
    ") -> tuple[list, dict]:\n",
    "    \"\"\"\n",
    "    Process categorical features (i.e. Domain has multiple sub-categories) from UniProt annotations.\n",
    "\n",
    "    Args:\n",
    "        column_data: Raw feature data from UniProt\n",
    "        column_name: Name of the feature column\n",
    "        category_options: Set of valid category names\n",
    "        seq_len: Length of the protein sequence\n",
    "        current_index: Dictionary tracking current index for each category\n",
    "\n",
    "    Returns:\n",
    "        Tuple of (list of category indices vectors, updated current_index)\n",
    "    \"\"\"\n",
    "    category_indices = {\n",
    "        category_name: [False] * seq_len for category_name in category_options\n",
    "    }\n",
    "    if pd.isna(column_data):\n",
    "        return [\n",
    "            category_indices[category] for category in category_options\n",
    "        ], current_index\n",
    "\n",
    "    entries = column_data.split(f\"{column_name} \")[1:]\n",
    "    for entry in entries:\n",
    "        positions_in_entry = entry.split(\";\")[0]\n",
    "        entry_category = re.search(r'/note=\"([^\"]+)\"', entry)\n",
    "        if entry_category:\n",
    "            entry_category = entry_category.group(1).split(\";\")[0]\n",
    "\n",
    "            if entry_category not in category_options:\n",
    "                # skip this one\n",
    "                continue\n",
    "\n",
    "            indices_in_entry = []\n",
    "            try:\n",
    "                # Skip the undefined cases\n",
    "                if \":\" in positions_in_entry or \"?\" in positions_in_entry:\n",
    "                    continue\n",
    "\n",
    "                # Case where the position is a range\n",
    "                if \"..\" in positions_in_entry:\n",
    "                    start, end = positions_in_entry.split(\"..\")\n",
    "                    start = start.strip(\"<\")\n",
    "                    end = end.strip(\">\")\n",
    "                    indices_in_entry = range(int(start), int(end) + 1)\n",
    "                else:\n",
    "                    indices_in_entry = [int(positions_in_entry)]\n",
    "                for index in indices_in_entry:\n",
    "                    if 0 <= index - 1 < seq_len:\n",
    "                        if entry_category in category_options:\n",
    "                            category_indices[entry_category][index - 1] = current_index[\n",
    "                                entry_category\n",
    "                            ]\n",
    "                        category_indices[\"any\"][index - 1] = current_index[\"any\"]\n",
    "\n",
    "                current_index[entry_category] += 1\n",
    "                current_index[\"any\"] += 1\n",
    "\n",
    "            except Exception as e:\n",
    "                print(\n",
    "                    f\"Error processing binary column index {positions_in_entry} with seq length {seq_len}\"\n",
    "                )\n",
    "    return [category_indices[category] for category in category_options], current_index\n",
    "\n",
    "\n",
    "def analyze_categorical_features(\n",
    "    df: pd.DataFrame, category: str, category_name: str, separator_name: str = \"note\"\n",
    ") -> Tuple[int, pd.Series, List[int], pd.Series]:\n",
    "    \"\"\"\n",
    "    Helper function for analyzing categorical features to find common categories and their statistics.\n",
    "\n",
    "    Returns:\n",
    "        Tuple of (number of proteins, occurrences per protein, lengths, note counts)\n",
    "    \"\"\"\n",
    "    non_na = df[category].dropna()\n",
    "    n_per_prot = non_na.apply(lambda x: x.count(category_name)).value_counts()\n",
    "\n",
    "    all_notes = []\n",
    "    all_lengths = []\n",
    "\n",
    "    for row in non_na:\n",
    "        entries = row.split(category_name)[1:]\n",
    "        for entry in entries:\n",
    "            # Extract note\n",
    "            note_start = entry.find(separator_name)\n",
    "            note = entry[note_start + len(separator_name) + 1 :]\n",
    "            note_end = note.find(\";\")\n",
    "            note = note[:note_end].strip('\"')\n",
    "\n",
    "            if \"/evidence\" not in note:\n",
    "                all_notes.append(note)\n",
    "\n",
    "            try:\n",
    "                # Extract length\n",
    "                location = entry[1:]\n",
    "                end_of_loc = location.find(\";\")\n",
    "                location = location[:end_of_loc] if end_of_loc != -1 else location\n",
    "\n",
    "                if \":\" in location or \"?\" in location:\n",
    "                    continue\n",
    "\n",
    "                if \"..\" in location:\n",
    "                    start, end = location.split(\"..\")\n",
    "                    start = int(start.strip(\"<\"))\n",
    "                    end = int(end.strip(\">\"))\n",
    "                    all_lengths.append(end - start)\n",
    "                else:\n",
    "                    all_lengths.append(1)\n",
    "\n",
    "            except Exception as e:\n",
    "                logger.warning(f\"Error analyzing feature length: {e}\")\n",
    "\n",
    "    return (\n",
    "        len(non_na),\n",
    "        n_per_prot.sort_values(ascending=False),\n",
    "        all_lengths,\n",
    "        pd.Series(all_notes).value_counts(),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b13adb99",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
